<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Sebastian Lobentanzer" />
  <meta name="author" content="Patrick Aloy" />
  <meta name="author" content="Jan Baumbach" />
  <meta name="author" content="Balazs Bohar" />
  <meta name="author" content="Vincent J. Carey" />
  <meta name="author" content="Pornpimol Charoentong" />
  <meta name="author" content="Katharina Danhauser" />
  <meta name="author" content="Tunca Doğan" />
  <meta name="author" content="Johann Dreo" />
  <meta name="author" content="Ian Dunham" />
  <meta name="author" content="Elias Farr" />
  <meta name="author" content="Adrià Fernandez-Torras" />
  <meta name="author" content="Benjamin M. Gyori" />
  <meta name="author" content="Michael Hartung" />
  <meta name="author" content="Charles Tapley Hoyt" />
  <meta name="author" content="Christoph Klein" />
  <meta name="author" content="Tamas Korcsmaros" />
  <meta name="author" content="Andreas Maier" />
  <meta name="author" content="Matthias Mann" />
  <meta name="author" content="David Ochoa" />
  <meta name="author" content="Elena Pareja-Lorente" />
  <meta name="author" content="Ferdinand Popp" />
  <meta name="author" content="Martin Preusse" />
  <meta name="author" content="Niklas Probul" />
  <meta name="author" content="Benno Schwikowski" />
  <meta name="author" content="Bünyamin Sen" />
  <meta name="author" content="Maximilian T. Strauss" />
  <meta name="author" content="Denes Turei" />
  <meta name="author" content="Erva Ulusoy" />
  <meta name="author" content="Dagmar Waltemath" />
  <meta name="author" content="Judith A. H. Wodke" />
  <meta name="author" content="Julio Saez-Rodriguez" />
  <meta name="dcterms.date" content="2023-11-22" />
  <meta name="keywords" content="markdown, publishing, manubot" />
  <title>Democratizing knowledge representation with BioCypher</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta property="og:type" content="article" />
  <meta name="dc.title" content="Democratizing knowledge representation with BioCypher" />
  <meta name="citation_title" content="Democratizing knowledge representation with BioCypher" />
  <meta property="og:title" content="Democratizing knowledge representation with BioCypher" />
  <meta property="twitter:title" content="Democratizing knowledge representation with BioCypher" />
  <meta name="dc.date" content="2023-11-22" />
  <meta name="citation_publication_date" content="2023-11-22" />
  <meta property="article:published_time" content="2023-11-22" />
  <meta name="dc.modified" content="2023-11-22T16:30:17+00:00" />
  <meta property="article:modified_time" content="2023-11-22T16:30:17+00:00" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Sebastian Lobentanzer" />
  <meta name="citation_author_institution" content="Heidelberg University, Faculty of Medicine and Heidelberg University Hospital, Institute for Computational Biomedicine, Heidelberg, Germany" />
  <meta name="citation_author_orcid" content="0000-0003-3399-6695" />
  <meta name="twitter:creator" content="@slobentanzer" />
  <meta name="citation_author" content="Patrick Aloy" />
  <meta name="citation_author_institution" content="Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain" />
  <meta name="citation_author_institution" content="Institució Catalana de Recerca i Estudis Avançats (ICREA), Barcelona, Catalonia, Spain" />
  <meta name="citation_author" content="Jan Baumbach" />
  <meta name="citation_author_institution" content="Institute for Computational Systems Biology, University of Hamburg, Germany" />
  <meta name="citation_author" content="Balazs Bohar" />
  <meta name="citation_author_institution" content="Earlham Institute, Norwich, UK" />
  <meta name="citation_author_institution" content="Biological Research Centre, Szeged, Hungary" />
  <meta name="citation_author" content="Vincent J. Carey" />
  <meta name="citation_author_institution" content="Channing Division of Network Medicine, Mass General Brigham, Harvard Medical School, Boston, USA" />
  <meta name="citation_author" content="Pornpimol Charoentong" />
  <meta name="citation_author_institution" content="Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany" />
  <meta name="citation_author_institution" content="Department of Medical Oncology, National Centre for Tumour Diseases (NCT), Heidelberg University Hospital (UKHD), Im Neuenheimer Feld 460, 69120, Heidelberg, Germany" />
  <meta name="citation_author" content="Katharina Danhauser" />
  <meta name="citation_author_institution" content="Department of Pediatrics, Dr. von Hauner Children’s Hospital, University Hospital, LMU Munich, Germany" />
  <meta name="citation_author" content="Tunca Doğan" />
  <meta name="citation_author_institution" content="Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author_institution" content="Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author" content="Johann Dreo" />
  <meta name="citation_author_institution" content="Computational Systems Biomedicine Lab, Department of Computational Biology, Institut Pasteur, Université Paris Cité, Paris, France" />
  <meta name="citation_author_institution" content="Bioinformatics and Biostatistics Hub, Institut Pasteur, Université Paris Cité, Paris, France" />
  <meta name="citation_author" content="Ian Dunham" />
  <meta name="citation_author_institution" content="European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK" />
  <meta name="citation_author_institution" content="Open Targets, Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK" />
  <meta name="citation_author" content="Elias Farr" />
  <meta name="citation_author_institution" content="Heidelberg University, Faculty of Medicine, and Heidelberg University Hospital, Institute for Computational Biomedicine, Bioquant, Heidelberg, Germany" />
  <meta name="citation_author" content="Adrià Fernandez-Torras" />
  <meta name="citation_author_institution" content="Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain" />
  <meta name="citation_author" content="Benjamin M. Gyori" />
  <meta name="citation_author_institution" content="Laboratory of Systems Pharmacology, Harvard Medical School, Boston, USA" />
  <meta name="citation_author" content="Michael Hartung" />
  <meta name="citation_author_institution" content="Institute for Computational Systems Biology, University of Hamburg, Germany" />
  <meta name="citation_author" content="Charles Tapley Hoyt" />
  <meta name="citation_author_institution" content="Laboratory of Systems Pharmacology, Harvard Medical School, Boston, USA" />
  <meta name="citation_author" content="Christoph Klein" />
  <meta name="citation_author_institution" content="Department of Pediatrics, Dr. von Hauner Children’s Hospital, University Hospital, LMU Munich, Germany" />
  <meta name="citation_author" content="Tamas Korcsmaros" />
  <meta name="citation_author_institution" content="Earlham Institute, Norwich, UK" />
  <meta name="citation_author_institution" content="Quadram Institute Bioscience, Norwich, UK" />
  <meta name="citation_author" content="Andreas Maier" />
  <meta name="citation_author_institution" content="Institute for Computational Systems Biology, University of Hamburg, Germany" />
  <meta name="citation_author" content="Matthias Mann" />
  <meta name="citation_author_institution" content="Proteomics Program, Novo Nordisk Foundation Centre for Protein Research, University of Copenhagen, Copenhagen, Denmark" />
  <meta name="citation_author_institution" content="Department of Proteomics and Signal Transduction, Max Planck Institute of Biochemistry, Martinsried, Germany" />
  <meta name="citation_author" content="David Ochoa" />
  <meta name="citation_author_institution" content="European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK" />
  <meta name="citation_author_institution" content="Open Targets, Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK" />
  <meta name="citation_author" content="Elena Pareja-Lorente" />
  <meta name="citation_author_institution" content="Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain" />
  <meta name="citation_author" content="Ferdinand Popp" />
  <meta name="citation_author_institution" content="Applied Tumour Immunity Clinical Cooperation Unit, National Centre for Tumour Diseases (NCT), German Cancer Research Centre (DKFZ), Im Neuenheimer Feld 460, 69120, Heidelberg, Germany" />
  <meta name="citation_author" content="Martin Preusse" />
  <meta name="citation_author_institution" content="Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany" />
  <meta name="citation_author" content="Niklas Probul" />
  <meta name="citation_author_institution" content="Institute for Computational Systems Biology, University of Hamburg, Germany" />
  <meta name="citation_author" content="Benno Schwikowski" />
  <meta name="citation_author_institution" content="Computational Systems Biomedicine Lab, Department of Computational Biology, Institut Pasteur, Université Paris Cité, Paris, France" />
  <meta name="citation_author" content="Bünyamin Sen" />
  <meta name="citation_author_institution" content="Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author_institution" content="Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author" content="Maximilian T. Strauss" />
  <meta name="citation_author_institution" content="Proteomics Program, Novo Nordisk Foundation Centre for Protein Research, University of Copenhagen, Copenhagen, Denmark" />
  <meta name="citation_author" content="Denes Turei" />
  <meta name="citation_author_institution" content="Heidelberg University, Faculty of Medicine, and Heidelberg University Hospital, Institute for Computational Biomedicine, Bioquant, Heidelberg, Germany" />
  <meta name="citation_author" content="Erva Ulusoy" />
  <meta name="citation_author_institution" content="Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author_institution" content="Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey" />
  <meta name="citation_author" content="Dagmar Waltemath" />
  <meta name="citation_author_institution" content="Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany" />
  <meta name="citation_author" content="Judith A. H. Wodke" />
  <meta name="citation_author_institution" content="Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany" />
  <meta name="citation_author" content="Julio Saez-Rodriguez" />
  <meta name="citation_author_institution" content="Heidelberg University, Faculty of Medicine and Heidelberg University Hospital, Institute for Computational Biomedicine, Heidelberg, Germany" />
  <meta name="citation_author_orcid" content="0000-0002-8552-8976" />
  <meta name="twitter:creator" content="@saezlab" />
  <link rel="canonical" href="https://biocypher.github.io/biocypher-paper/" />
  <meta property="og:url" content="https://biocypher.github.io/biocypher-paper/" />
  <meta property="twitter:url" content="https://biocypher.github.io/biocypher-paper/" />
  <meta name="citation_fulltext_html_url" content="https://biocypher.github.io/biocypher-paper/" />
  <meta name="citation_pdf_url" content="https://biocypher.github.io/biocypher-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://biocypher.github.io/biocypher-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://biocypher.github.io/biocypher-paper/v/c4ef697a550e8ca4950a3641b22508988a354771/" />
  <meta name="manubot_html_url_versioned" content="https://biocypher.github.io/biocypher-paper/v/c4ef697a550e8ca4950a3641b22508988a354771/" />
  <meta name="manubot_pdf_url_versioned" content="https://biocypher.github.io/biocypher-paper/v/c4ef697a550e8ca4950a3641b22508988a354771/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Democratizing knowledge representation with BioCypher</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://biocypher.github.io/biocypher-paper/v/c4ef697a550e8ca4950a3641b22508988a354771/">permalink</a>)
was automatically generated
from <a href="https://github.com/biocypher/biocypher-paper/tree/c4ef697a550e8ca4950a3641b22508988a354771">biocypher/biocypher-paper@c4ef697</a>
on November 22, 2023.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Sebastian Lobentanzer</strong>
<sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-3399-6695">0000-0003-3399-6695</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/slobentanzer">slobentanzer</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/slobentanzer">slobentanzer</a>
<br>
<small>
Heidelberg University, Faculty of Medicine and Heidelberg University Hospital, Institute for Computational Biomedicine, Heidelberg, Germany
</small></p></li>
<li><p><strong>Patrick Aloy</strong>
<br>
<small>
Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain; Institució Catalana de Recerca i Estudis Avançats (ICREA), Barcelona, Catalonia, Spain
</small></p></li>
<li><p><strong>Jan Baumbach</strong>
<br>
<small>
Institute for Computational Systems Biology, University of Hamburg, Germany
</small></p></li>
<li><p><strong>Balazs Bohar</strong>
<br>
<small>
Earlham Institute, Norwich, UK; Biological Research Centre, Szeged, Hungary
</small></p></li>
<li><p><strong>Vincent J. Carey</strong>
<br>
<small>
Channing Division of Network Medicine, Mass General Brigham, Harvard Medical School, Boston, USA
</small></p></li>
<li><p><strong>Pornpimol Charoentong</strong>
<br>
<small>
Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany; Department of Medical Oncology, National Centre for Tumour Diseases (NCT), Heidelberg University Hospital (UKHD), Im Neuenheimer Feld 460, 69120, Heidelberg, Germany
</small></p></li>
<li><p><strong>Katharina Danhauser</strong>
<br>
<small>
Department of Pediatrics, Dr. von Hauner Children’s Hospital, University Hospital, LMU Munich, Germany
</small></p></li>
<li><p><strong>Tunca Doğan</strong>
<br>
<small>
Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey; Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey
</small></p></li>
<li><p><strong>Johann Dreo</strong>
<br>
<small>
Computational Systems Biomedicine Lab, Department of Computational Biology, Institut Pasteur, Université Paris Cité, Paris, France; Bioinformatics and Biostatistics Hub, Institut Pasteur, Université Paris Cité, Paris, France
</small></p></li>
<li><p><strong>Ian Dunham</strong>
<br>
<small>
European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK; Open Targets, Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK
</small></p></li>
<li><p><strong>Elias Farr</strong>
<br>
<small>
Heidelberg University, Faculty of Medicine, and Heidelberg University Hospital, Institute for Computational Biomedicine, Bioquant, Heidelberg, Germany
</small></p></li>
<li><p><strong>Adrià Fernandez-Torras</strong>
<br>
<small>
Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain
</small></p></li>
<li><p><strong>Benjamin M. Gyori</strong>
<br>
<small>
Laboratory of Systems Pharmacology, Harvard Medical School, Boston, USA
</small></p></li>
<li><p><strong>Michael Hartung</strong>
<br>
<small>
Institute for Computational Systems Biology, University of Hamburg, Germany
</small></p></li>
<li><p><strong>Charles Tapley Hoyt</strong>
<br>
<small>
Laboratory of Systems Pharmacology, Harvard Medical School, Boston, USA
</small></p></li>
<li><p><strong>Christoph Klein</strong>
<br>
<small>
Department of Pediatrics, Dr. von Hauner Children’s Hospital, University Hospital, LMU Munich, Germany
</small></p></li>
<li><p><strong>Tamas Korcsmaros</strong>
<br>
<small>
Earlham Institute, Norwich, UK; Quadram Institute Bioscience, Norwich, UK
</small></p></li>
<li><p><strong>Andreas Maier</strong>
<br>
<small>
Institute for Computational Systems Biology, University of Hamburg, Germany
</small></p></li>
<li><p><strong>Matthias Mann</strong>
<br>
<small>
Proteomics Program, Novo Nordisk Foundation Centre for Protein Research, University of Copenhagen, Copenhagen, Denmark; Department of Proteomics and Signal Transduction, Max Planck Institute of Biochemistry, Martinsried, Germany
</small></p></li>
<li><p><strong>David Ochoa</strong>
<br>
<small>
European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK; Open Targets, Wellcome Genome Campus, Hinxton, Cambridgeshire CB10 1SD, UK
</small></p></li>
<li><p><strong>Elena Pareja-Lorente</strong>
<br>
<small>
Institute for Research in Biomedicine (IRB Barcelona), the Barcelona Institute of Science and Technology, Barcelona, Catalonia, Spain
</small></p></li>
<li><p><strong>Ferdinand Popp</strong>
<br>
<small>
Applied Tumour Immunity Clinical Cooperation Unit, National Centre for Tumour Diseases (NCT), German Cancer Research Centre (DKFZ), Im Neuenheimer Feld 460, 69120, Heidelberg, Germany
</small></p></li>
<li><p><strong>Martin Preusse</strong>
<br>
<small>
Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany
</small></p></li>
<li><p><strong>Niklas Probul</strong>
<br>
<small>
Institute for Computational Systems Biology, University of Hamburg, Germany
</small></p></li>
<li><p><strong>Benno Schwikowski</strong>
<br>
<small>
Computational Systems Biomedicine Lab, Department of Computational Biology, Institut Pasteur, Université Paris Cité, Paris, France
</small></p></li>
<li><p><strong>Bünyamin Sen</strong>
<br>
<small>
Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey; Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey
</small></p></li>
<li><p><strong>Maximilian T. Strauss</strong>
<br>
<small>
Proteomics Program, Novo Nordisk Foundation Centre for Protein Research, University of Copenhagen, Copenhagen, Denmark
</small></p></li>
<li><p><strong>Denes Turei</strong>
<br>
<small>
Heidelberg University, Faculty of Medicine, and Heidelberg University Hospital, Institute for Computational Biomedicine, Bioquant, Heidelberg, Germany
</small></p></li>
<li><p><strong>Erva Ulusoy</strong>
<br>
<small>
Biological Data Science Lab, Department of Computer Engineering, Hacettepe University, Ankara, Turkey; Department of Bioinformatics, Graduate School of Health Sciences, Hacettepe University, Ankara, Turkey
</small></p></li>
<li><p><strong>Dagmar Waltemath</strong>
<br>
<small>
Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany
</small></p></li>
<li><p><strong>Judith A. H. Wodke</strong>
<br>
<small>
Centre for Quantitative Analysis of Molecular and Cellular Biosystems (Bioquant), Heidelberg University, Im Neuenheimer Feld 267, 69120, Heidelberg, Germany
</small></p></li>
<li><p><strong>Julio Saez-Rodriguez</strong>
<sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8552-8976">0000-0002-8552-8976</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/saezrodriguez">saezrodriguez</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/saezlab">saezlab</a>
<br>
<small>
Heidelberg University, Faculty of Medicine and Heidelberg University Hospital, Institute for Computational Biomedicine, Heidelberg, Germany
</small></p></li>
</ul>
<div id="correspondence">
<p>✉ — Correspondence possible via <a href="https://github.com/biocypher/biocypher-paper/issues">GitHub Issues</a>
or email to
Sebastian Lobentanzer &lt;sebastian.lobentanzer@gmail.com&gt;,
Julio Saez-Rodriguez &lt;pub.saez@uni-heidelberg.de&gt;.</p>
</div>
<h2 class="page_break_before" id="main">Main</h2>
<p>Biomedical data are amassed at an ever-increasing rate, and machine learning tools that use prior knowledge in combination with biomedical big data are gaining much traction <span class="citation" data-cites="rf9J6yvQ vhsYX2y">[<a href="#ref-rf9J6yvQ" role="doc-biblioref">1</a>,<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>.
Knowledge graphs (KGs) are rapidly becoming the dominant form of knowledge representation.
KGs are data structures that represent knowledge as a graph to facilitate navigation and analysis of complex information, often by leveraging semantic information.
Their versatility has made them popular in areas such as data storage, reasoning, and explainable artificial intelligence <span class="citation" data-cites="CiGapf55">[<a href="#ref-CiGapf55" role="doc-biblioref">3</a>]</span>.
However, for many research groups, building their own biomedical KG is prohibitively expensive.
This motivated us to build the BioCypher framework to support users in creating KGs (https://biocypher.org).</p>
<p>The ability to build a task-specific KG is important, since directly standardising the representation of biomedical knowledge is not appropriate for the diverse research tasks in the community.
While human researchers can contextualise and abstract concepts easily, the same does not apply to algorithms.
For example, drug discovery tasks (viewing genes as functional ancestors of protein targets) require a different KG structure and content compared to the implementation of a molecular tumour board (genes as clinical markers), which is different still from research into cell type-contextualised gene regulatory network inference (genes as targets of regulatory mechanisms).
Even for similar tasks, the KG structure or subtle decisions about included resources lead to different results for many modern analytic methods <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>.
In addition, decisions about how to represent knowledge at each primary resource pose problems in their integration, for instance via the use of different identifier namespaces, levels of granularity, or licences <span class="citation" data-cites="yqoUY2f9 TNqtTxS8">[<a href="#ref-yqoUY2f9" role="doc-biblioref">4</a>,<a href="#ref-TNqtTxS8" role="doc-biblioref">5</a>]</span>.</p>
<p>The current landscape of biomedical KGs is not easily navigated; neither the KGs themselves, nor the pipelines used to build them, consistently adhere to FAIR (Findable, Accessible, Interoperable, and Reusable) <span class="citation" data-cites="6DjakjNS">[<a href="#ref-6DjakjNS" role="doc-biblioref">6</a>]</span> and TRUST (Transparency, Responsibility, User focus, Sustainability, and Technology) <span class="citation" data-cites="MMp6cQ1x">[<a href="#ref-MMp6cQ1x" role="doc-biblioref">7</a>]</span> principles.
Understandably, the overhead required to implement these principles may not be justified when building a one-off task-specific KG for research.
Thus, many KGs are built manually for specific applications, which leads to issues in their reuse and integration <span class="citation" data-cites="yqoUY2f9">[<a href="#ref-yqoUY2f9" role="doc-biblioref">4</a>]</span>.
For downstream users, the resulting KGs are too distinct to easily compare or combine <span class="citation" data-cites="TNqtTxS8">[<a href="#ref-TNqtTxS8" role="doc-biblioref">5</a>]</span>.
Maintaining KGs for the community is additional work; once maintenance stops, they quickly deteriorate, leading to reusability and reproducibility issues <span class="citation" data-cites="yqoUY2f9">[<a href="#ref-yqoUY2f9" role="doc-biblioref">4</a>]</span> (<a href="#supplementary-note-1---background">Supplementary Note 1</a>).</p>
<p>BioCypher has been built with continuous consideration of the FAIR and TRUST principles, yielding benefits to the entire community in multiple respects:</p>
<p><strong>Modularity:</strong> To rationalise efforts across the community, we propose a modular architecture that maximises reuse of data and code in three ways: input, ontology, and output (Figure 1A).
Input adapters allow delegating maintenance work to one central place for each resource, ontology adapters give access to the wealth of structured information curated by the ontology community, and output adapters allow benchmarking and selection of database management systems.
Together, these mechanisms enable a workflow that reduces the time and effort to develop and deploy custom KGs.</p>
<p><strong>Harmonisation:</strong> By using ontologies as expertly crafted repositories of conceptual hierarchies, we facilitate harmonisation from a biological perspective.
We aid with the technical aspects of using and manipulating ontologies, for instance by flexibly extending or hybridising complementary ontologies.</p>
<p><strong>Reproducibility:</strong> By sharing the mapping of KG contents to ontologies, we facilitate reproduction of the structure of the corresponding database without access to the primary data, which may be prohibited by licence or privacy issues.
We also enable extraction of subgraphs, effectively converting storage-oriented to task-specific KGs, which due to their reduced sizes are easier to share alongside analyses.</p>
<p><strong>Reusability and accessibility:</strong> Finally, the sustainability of research software is strongly related to adoption in – and contributions from – the community.
BioCypher is developed as a TRUSTworthy open-source software, applying methods of continuous integration and deployment, and including a diverse community of researchers and developers from the beginning.
This facilitates workflows that are tested end-to-end, including the integrity of the scientific data.
We operate under the permissive MIT licence and provide community members with guidelines for their contributions and a code of conduct (https://github.com/biocypher).</p>
<p>Different measures further increase the accessibility and FAIRness of our framework.
For example, we provide a template repository for a BioCypher pipeline with adapters, including a Docker Compose setup.
To enable learning by example, we curate existing pipelines, as well as all adapters they use, in our GitHub organisation.
Using the GitHub API and a BioCypher pipeline, we build a “meta-graph” for the simple browsing and analysis of BioCypher workflows (https://meta.biocypher.org).
To inform the contents of this meta-graph, we have reactivated and now maintain the Biomedical Resource Ontology (BRO <span class="citation" data-cites="1ALlMCWlM">[<a href="#ref-1ALlMCWlM" role="doc-biblioref">8</a>]</span>), which helps to categorise pipelines and adapters into research areas, data types, and purposes (<a href="#supplementary-note-2---approach">Supplementary Note 2</a>).</p>
<p><img src="https://biocypher.org/_images/graphical_abstract.png" alt="Figure 1: The BioCypher framework. A) Threefold modularity: Resources (left) and ontologies (bottom left) are combined to yield a knowledge graph (right). The mapping of entities to ontology concepts is realised by shareable configuration, which can be iteratively optimised. B) Initially, we transform commonly used, curated “secondary” resources into configurable, task-specific knowledge graphs in various output formats. Incrementally, these secondary adapters will be replaced by primary resource adapters (see Figure S1). Coloured panels in A and B indicate parts of the BioCypher ecosystem. C) Agreeing on a common representational framework allows harmonisation of task-specific data sources to answer complex queries across biomedical domains. For instance, starting at mass spectrometry measurements of a patient’s tumour (left), one could go through clinical annotations to genetic dependencies from the Dependency Map project to identify potential drug targets, or through pathway / process annotations in Reactome and IntAct, identify relevant ligand-receptor pairs using OmniPath, and use CROssBAR to perform drug discovery or repurposing for these receptors. Panels correspond to resources; although we work on most of the displayed resources, the figure is used for illustrative purposes and does not depict an existing pipeline." />
{#fig:1}</p>
<p>BioCypher is implemented as a Python library that provides a low-code access point to data processing and ontology manipulation, emphasising the reuse of existing resources to the highest extent possible.
We have begun to open the platform to other bioinformatics ecosystems, starting with R/Bioconductor (https://biocypher.org/r-bioc.html).
By our design principles and the automation of data management tasks, we aim to free up developer time and guide decision making on how to represent knowledge, bridging the gap between the field of biomedical ontology and the broad application of databases in research.</p>
<p>By abstracting the KG build process as a combination of modular input adapters, we save developer time in the maintenance of integrative resources built from overlapping primary sources (Figure 1B), for instance OmniPath <span class="citation" data-cites="BDzI9Y5F">[<a href="#ref-BDzI9Y5F" role="doc-biblioref">9</a>]</span>, Bioteque <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>, CROssBAR DB <span class="citation" data-cites="57ZT8woI">[<a href="#ref-57ZT8woI" role="doc-biblioref">10</a>]</span>, and the Clinical Knowledge Graph <span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span>.
By mapping the contents of those resources onto a common ontological space, we gain interoperability between the different biomedical domains (Figure 1C).
BioCypher helps with the mapping procedure by providing examples and an interface, as well as numerous user-friendliness measures.
By using the industry standard Web Ontology Language (OWL) format, we provide access to the majority of available ontologies.
Separating the ontology framework from the modelled data enables the implementation of reasoning applications at the ontology level, for instance the ad-hoc harmonisation of disease ontologies.</p>
<p>By providing access to a range of modular output adapters, we facilitate the project-specific benchmarking and selection of suitable database management systems.
For instance, a Neo4j adapter provides rapid access to extensive databases for maintenance of knowledge and enables queries from analysis (Jupyter) notebooks.
Switching to alternative graph or relational databases (e.g., ArangoDB or PostgreSQL) allows for task-specific performance optimisation.
A CSV-writer and Python-native adapters (e.g., Pandas, sparse matrix, or NetworkX formats) yield knowledge representations that can directly be used programmatically by a wide range of machine learning frameworks.
Due to BioCypher’s modular nature, additional output adapters can quickly be added.</p>
<p>Application programming interfaces (APIs) built on top of the BioCypher KGs enable complex and versatile queries and simplify the interaction of users with the knowledge.
For example, web widgets and apps (such as drug discovery and repositioning with https://crossbar.kansil.org and analysis workflows with https://drugst.one) allow researchers to browse and customise the database, and to plug it into standard pipelines.
Additionally, a structured, semantically enriched knowledge representation facilitates connection to and improves performance of modern natural language processing applications such as GPT <span class="citation" data-cites="erUdG5Ur">[<a href="#ref-erUdG5Ur" role="doc-biblioref">12</a>]</span>, which can be specifically tuned for biomedical research <span class="citation" data-cites="11YmMTmFW">[<a href="#ref-11YmMTmFW" role="doc-biblioref">13</a>]</span>.
The use of common standards enables sharing of tools across projects and communities or in cloud-based services that preserve sensitive patient data (<a href="#supplementary-note-3---implementation">Supplementary Note 3</a>).</p>
<p>There have been numerous attempts at standardising KGs and making biomedical data stores more interoperable.
We can identify three general types of approaches, in increasing order of abstraction: centrally maintained databases, explicit standard formats (modelling languages), and KG frameworks.
With BioCypher, we aim to improve user-friendliness on all three levels of abstraction; for an in-depth discussion, see <a href="#supplementary-note-4---prior-art">Supplementary Note 4</a>.
Despite many efforts, there is no widely accepted solution.
Very often, resources take the “path of least resistance” in adopting their own, arbitrary formats of representation. To our knowledge, no framework provides easy access to state-of-the-art KGs to the average biomedical researcher, a gap that BioCypher aims to fill.
We demonstrate some key advantages of BioCypher by case studies in <a href="#supplementary-note-5---case-studies">Supplementary Note 5</a>.</p>
<p>We believe that creating a more interoperable biomedical research community is as much a social effort as it is a scientific software problem.
To facilitate adoption of any approach, the process must be made as simple as possible, and it must yield tangible rewards, such as significant savings in developer time.
We will provide hands-on training for all interested researchers, and we invite all database and tool developers to join our collective effort.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>This project has received funding from the European Union’s Horizon 2020 research and innovation programme (grant agreement No 965193 [DECIDER] and 116030 [TransQST]), the German Federal Ministry of Education and Research (BMBF, Computational Life Sciences grant No 031L0181B and MSCoreSys research initiative research core SMART-CARE 031L0212A), the Defense Advanced Research Projects Agency (DARPA) Young Faculty Award [W911NF-20-1-0255], and the Medical Informatics Initiative Germany, MIRACUM consortium, FKZ: 01ZZ2019.</p>
<p>We thank Henning Hermjakob, Benjamin Haibe-Kains, Pablo Rodriguez-Mier, Daniel Dimitrov, and Olga Ivanova for feedback on the manuscript, and Ben Hitz and Pedro Assis for feedback on their use of BioCypher.</p>
<h2 id="author-contributions">Author Contributions</h2>
<p>The project was conceived by SL and JSR.
The software was developed by SL with input from DT.
The manuscript was drafted by SL, edited by JSR, and jointly revised by all co-authors.
All co-authors as members of the BioCypher Consortium contributed to the case studies in development and writing and gave feedback for software development, which was coordinated and integrated by SL.</p>
<h2 id="conflict-of-interest">Conflict of Interest</h2>
<p>JSR reports funding from GSK, Pfizer and Sanofi and fees from Travere Therapeutics and Astex Pharmaceuticals.</p>
<h2 id="supplementary-methods">Supplementary Methods</h2>
<p>BioCypher is implemented as a Python package.
Its structure follows the purpose of a threefold modularity of inputs, ontology, and outputs.
A user interface class (“core”) receives user choices via configuration YAML files and connects the inputs provided by resource-specific adapters to either bulk disk-writing methods or driver-based connections tailor-made for database management systems.
It also manages the mapping of data inputs to ontologies with the help of an ontology module.
This modular architecture facilitates extension of all modules according to the community’s needs.</p>
<p>The resulting knowledge graphs (KGs) can be described as “instance-based” realisations of biomedical concepts: using the concept definition from the ontology, each entity in the graph becomes an instance of this concept.
We recommend the use of a generic “high-level” ontology such as the Biolink model <span class="citation" data-cites="vzPcrJo9">[<a href="#ref-vzPcrJo9" role="doc-biblioref">14</a>]</span>, a comprehensive and generic biomedical ontology; where needed, this ontology can be exchanged with or extended by more specific and task-directed ontologies, for instance from the OBO Foundry <span class="citation" data-cites="CNP39tah">[<a href="#ref-CNP39tah" role="doc-biblioref">15</a>]</span>.
The versions of all used ontologies should be specified by each pipeline, which can most effectively be realised by specifying a persistent URL (PURL) for the versioned ontology file (most commonly in OWL format) in the BioCypher configuration.
Identifier namespaces are collected from the community-curated and frequently updated Bioregistry service <span class="citation" data-cites="fKdMnXba">[<a href="#ref-fKdMnXba" role="doc-biblioref">16</a>]</span>, which is important for ensuring continued compatibility among the created KGs.
Bioregistry also supplies convenient methods for parsing identifier Compact URIs (CURIEs), which are the preferred method of unambiguously specifying identities of KG entities.
For identifier mapping, where required, the corresponding facilities of pypath <span class="citation" data-cites="BDzI9Y5F">[<a href="#ref-BDzI9Y5F" role="doc-biblioref">9</a>]</span> are used and extended.</p>
<p>The preferred way of entering data into a BioCypher graph attaches scientific provenance to each entry, allowing the aggregation of data with respect to their sources (for instance, the publication an interaction was derived from) and thus avoiding problems such as duplicate counting of the same primary data from different secondary curations.
For author attribution, the preferred way of entering data into BioCypher also includes the exact provenance of each entry.
In the same way, all licences of the contents are propagated forward, enabling the users of the framework to easily determine the allowed uses for any given KG.
This behaviour can be enforced by using BioCypher’s “strict mode.” The attachment of this information can be particularly useful in cases in which a subset of the graph does not fulfil the user’s requirements; individual entity annotation allows the usage of only the parts of the KG that are covered by the rights of the user.
While the ultimate responsibility of correct interpretation and execution of licensing issues lies with the end user, we strive to make the task as accessible as possible.</p>
<p>BioCypher is a free software under MIT licence, openly developed and available at https://github.com/biocypher and via PyPI.
We are generally compatible with the three most recent Python versions (which currently is 3.9 or higher).
Community contributions in the form of GitHub issues or pull requests are very welcome and encouraged.
More details and a tutorial can be found in the documentation at https://biocypher.org.</p>
<h2 id="supplementary-note-1---background">Supplementary Note 1 - Background</h2>
<p>We here give some background and references on the problem of standardising biomedical knowledge representation.
Biomedical knowledge, although increasingly abundant, is fragmented across hundreds of resources.
For instance, a clinical researcher may use protein information from UniProtKB <span class="citation" data-cites="cvbuQhIz">[<a href="#ref-cvbuQhIz" role="doc-biblioref">17</a>]</span>, genetic variants from COSMIC <span class="citation" data-cites="ezOoICHE">[<a href="#ref-ezOoICHE" role="doc-biblioref">18</a>]</span>, protein interactions from IntAct <span class="citation" data-cites="dGRWxzm4">[<a href="#ref-dGRWxzm4" role="doc-biblioref">19</a>]</span>, and information on clinical trials from ClinicalTrials.gov <span class="citation" data-cites="NQaiGeFN">[<a href="#ref-NQaiGeFN" role="doc-biblioref">20</a>]</span>.</p>
<p>Finding the most suitable KG for a specific task is challenging and time-consuming; they are published in isolation and there is no registry <span class="citation" data-cites="yqoUY2f9 TNqtTxS8">[<a href="#ref-yqoUY2f9" role="doc-biblioref">4</a>,<a href="#ref-TNqtTxS8" role="doc-biblioref">5</a>]</span>.
Few available KG solutions perfectly fit the task the individual researcher wants to perform, but creating custom KGs is only possible for those that can afford years of development time by an individual <span class="citation" data-cites="1GEdqX9gO vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>,<a href="#ref-1GEdqX9gO" role="doc-biblioref">21</a>]</span> or even entire teams <span class="citation" data-cites="SmbvHVqc">[<a href="#ref-SmbvHVqc" role="doc-biblioref">22</a>]</span>.
Smaller or non-bioinformatics labs need to choose from publicly available KGs, limiting customisation and the use of non-public data.
There exist frameworks to build certain kinds of KG from scratch <span class="citation" data-cites="1H2nqqKV7 mZ5zqhmN">[<a href="#ref-1H2nqqKV7" role="doc-biblioref">23</a>,<a href="#ref-mZ5zqhmN" role="doc-biblioref">24</a>]</span>, but these are difficult to use for researchers outside of the ontology subfield and often have a rigid underlying data model 10,<span class="citation" data-cites="1353itlbv">[<a href="#ref-1353itlbv" role="doc-biblioref">25</a>]</span>.
Even task-specific knowledge graphs sometimes need to be built locally by the user due to licensing or maintenance reasons, which requires significant technical expertise <span class="citation" data-cites="pyC7TREj">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>]</span>.
Modifying an existing, comprehensive KG for a specific purpose is a non-trivial and often manual process prone to lack of reproducibility <span class="citation" data-cites="DeIC0Dgw">[<a href="#ref-DeIC0Dgw" role="doc-biblioref">27</a>]</span>.</p>
<h2 id="supplementary-note-2---approach">Supplementary Note 2 - Approach</h2>
<p>We expand here on our section in the main text, detailing the four pillars of our approach.</p>
<ol type="1">
<li><p>Modularity: To rationalise efforts across the community, we propose a flexible modular architecture that maximises reuse of data and code in three ways: input, ontology, and output.
Input adapters allow delegating maintenance work to one central place for each resource, ontology adapters give access to the wealth of structured information curated by the ontology community, and output adapters allow benchmarking and selection of database management systems.
Together, these mechanisms shall enable an agile workflow that drastically reduces the time and effort to develop and deploy custom KGs for small teams.
See the case study “Modularity” and others for more information.</p></li>
<li><p>Harmonisation: To facilitate harmonisation of datasets from a biological perspective, we propose to use ontology mapping (referring to a hierarchical organisation of biological concepts).
Primary data sources may represent similar data in different ways.
BioCypher harmonises biomedical data by mapping divergent representations onto the same biological concept, and aids with the technical aspects of ontology manipulation (see case study “Tumour board”).
In addition, the ontological information projected onto each KG entity allows for more flexible and informative queries in downstream analyses (see case study “Network expansion”).</p></li>
<li><p>Reproducibility: By sharing the ontology mapping from (2) in a project-specific manner, a database used for a specific task can be reproduced more effectively.
Since sharing the databases themselves is often prohibited by their large size, BioCypher facilitates the creation of task-specific subsets of databases to be shared alongside analyses.
Extensive automation reduces development time and file sizes, while additionally making the shared dataset independent of database software versions (see case studies “Network expansion”, “Subgraph extraction”, and “Embedding”).</p></li>
<li><p>Reusability and accessibility: Our template repository for a BioCypher pipeline with adapters, including a Docker Compose setup, is available <a href="https://github.com/biocypher/project-template">on GitHub</a>.
To enable learning by example, we curate existing pipelines as well as all adapters they use in a <a href="https://github.com/orgs/biocypher/projects/3">GitHub project</a> that is tied to the BioCypher repository.
With these data, using the GitHub API and a pipeline based on our template, we build a BioCypher “meta-graph” for the simple browsing and analysis of existing BioCypher workflows (https://meta.biocypher.org/).
To inform the structure of this meta-graph, we have reactivated and now maintain the <a href="https://bioportal.bioontology.org/ontologies/BRO/?p=summary">Biomedical Resource Ontology</a> (BRO <span class="citation" data-cites="1ALlMCWlM">[<a href="#ref-1ALlMCWlM" role="doc-biblioref">8</a>]</span>), which allows the categorisation of pipelines and adapters (now <a href="https://github.com/biocypher/biomedical-resource-ontology">on GitHub</a>).</p></li>
</ol>
<p>While data FAIRness is a necessary part of open science communication, it is not sufficient for the adoption and sustainability of a software project such as BioCypher.
As such, we also implement measures based on the TRUST principles, to increase usability, accessibility, and extensibility of our framework.
For more information, see the following Supplementary Text on “Sustainable Development.”</p>
<h3 id="sustainable-development">Sustainable Development</h3>
<p>We have implemented numerous measures to increase the user-friendliness of our framework.
The BioCypher ecosystem is maintained centrally at https://github.com/biocypher, which includes projects for the management of development and the components of BioCypher pipelines (adapters and ontologies).
These projects serve as the ground truth for available BioCypher modules, and are used by a BioCypher pipeline (https://github.com/biocypher/meta-graph) to build an overview graph database that is automatically deployed to our server as a freely accessible Neo4j browser instance (at https://meta.biocypher.org, no login credentials required).
Prospective users can use the board and the graph to find examples and reusable components for their own KG.</p>
<p>We provide a template repository (https://github.com/biocypher/project-template) that guides new users through the process of deploying their own KG.
It includes a <code>docker compose</code> setup which can be used to execute the KG build step and automatically transfer the KG into a Neo4j database running in the official Neo4j Docker container, thus being automatically secure to deploy.</p>
<p>We provide a detailed tutorial for all aspects of BioCypher on our web page, https://biocypher.org, which we update regularly as new features are added.
We provide easy access to our community on that page, including email contact, a mailing list, and a community chat channel at https://biocypher.zulipchat.com.
We also explicitly encourage contributions and getting in contact, and we offer help through online or in-person seminars and meetings.
We provide community guidelines, a code of conduct, and a developer guide for contributing.
We participate in and organise hackathons to educate about knowledge representation and improve interoperability with other software ecosystems, such as Bioconductor and Galaxy.</p>
<h2 id="supplementary-note-3---implementation">Supplementary Note 3 - Implementation</h2>
<p>We build on recent technological and conceptual developments in biomedical ontologies that greatly facilitate the harmonisation of biomedical knowledge and advocate a philosophy of reuse of open-source software.
For instance, we integrate a comprehensive “high-level” biomedical ontology, the Biolink model 1, which can be replaced or extended by more domain-specific ontologies as needed, and an extensive catalogue and resolver for biomedical identifier resources, the Bioregistry 3.
Both projects, like BioCypher, are open-source and community-driven.
The ontologies serve as a framework for the representation of biomedical concepts; by supporting the Web Ontology Language (OWL), BioCypher allows integration and manipulation of most ontologies, including <a href="https://github.com/monarch-initiative/ontogpt">those generated by Large Language Models</a>.</p>
<p>Separating the ontology framework from the modelled data allows implementation of reasoning applications at the ontology level, for instance the ad-hoc harmonisation of multiple disease ontologies before mapping the data points.
For instance, with a group of users that are knowledgeable in ontology, a way to harmonise the divergent or incomplete ontologies can be developed, e.g.
on the topic of diseases, before using them to inform the knowledge representation output.
In addition, new developments in the field of language models and grounding will enable plugging “automatic” grounding into the ontology adapter in BioCypher, helping more novice users with the mapping between KG entities and the corresponding ontologies (see for instance https://github.com/ccb-hms/ontology-mapper).</p>
<p>Building a task-specific KG, given existing configuration, takes only minutes, and creating a KG from scratch can be achieved in a few days of work.
This allows for rapid prototyping and automated machine learning (ML) pipelines that iterate the KG structure to optimise predictive performance; for instance, building custom task-specific KGs for graph embeddings and ML (see case study “Embeddings”).
Despite its speed, automated testing of millions of entities and relationships per KG increases trust in the consistency of the data (see Supplementary Methods for details and the case study “Network expansion” for an example).</p>
<h2 id="supplementary-note-4---prior-art">Supplementary Note 4 - Prior Art</h2>
<p>There have been numerous attempts at standardising knowledge graphs and making biomedical data stores more interoperable <span class="citation" data-cites="yqoUY2f9 TNqtTxS8">[<a href="#ref-yqoUY2f9" role="doc-biblioref">4</a>,<a href="#ref-TNqtTxS8" role="doc-biblioref">5</a>]</span>.
They can be divided into three broad classes representing increasing levels of abstraction of the KG build process:</p>
<ol type="1">
<li><p>Centrally maintained databases include task-oriented data collections such as OmniPath 4 or the CKG <span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span>.
They are the least flexible form of knowledge representation, usually bound to a specific research purpose, and are highly dependent on their primary maintainers for continuous functioning.
BioCypher reduces the development and maintenance overhead that usually goes along with such a resource, making a task-specific KG feasible for smaller and less bioinformatics-focused groups.
These databases usually do not conform to any standard in their knowledge representation, hindering their integration.
In contrast, with BioCypher, we migrate OmniPath, CKG, and other popular databases onto an interoperable KG framework.</p></li>
<li><p>Explicit standard formats or modelling languages include the Biolink model 1, BEL <span class="citation" data-cites="11W3tHiHT">[<a href="#ref-11W3tHiHT" role="doc-biblioref">28</a>]</span>, GO-CAM <span class="citation" data-cites="akdxNohe">[<a href="#ref-akdxNohe" role="doc-biblioref">29</a>]</span>, SBML <span class="citation" data-cites="mP9wIAFk">[<a href="#ref-mP9wIAFk" role="doc-biblioref">30</a>]</span>, BioPAX <span class="citation" data-cites="1Cn9nrdiq">[<a href="#ref-1Cn9nrdiq" role="doc-biblioref">31</a>]</span>, and PSI-MI <span class="citation" data-cites="KwOAgLE2">[<a href="#ref-KwOAgLE2" role="doc-biblioref">32</a>]</span>.
There are many more, each a solution to a very specific problem, as reviewed elsewhere <span class="citation" data-cites="11W3tHiHT Nxf46J67">[<a href="#ref-11W3tHiHT" role="doc-biblioref">28</a>,<a href="#ref-Nxf46J67" role="doc-biblioref">33</a>]</span>; some are part of the COMBINE standard ecosystem <span class="citation" data-cites="PqRxavyS">[<a href="#ref-PqRxavyS" role="doc-biblioref">34</a>]</span>.
Their main shortcoming is the rigidity that follows from their data model definitions: to represent data in one of these languages, the user needs to fully adopt it.
If the task exceeds the scope of the language, the user needs to either look for alternatives, or introduce new features into the language, which can be a lengthy process.
In addition, some features may be incompatible, and thus, one centrally maintained language definition is fundamentally limited.
With BioCypher, each of the above languages can be adopted as the basis for a particular knowledge graph; in fact, we use the Biolink model as a basic ontology.
Inside our framework, these languages can be freely and transparently exchanged, modified, extended, and hybridised, as we show in several of our case studies (e.g., “Tumour board” extends Biolink with Sequence Ontology and Disease Ontology).</p></li>
<li><p>KG frameworks provide a means to build KGs, similar to the idea of BioCypher 14;<span class="citation" data-cites="mZ5zqhmN">[<a href="#ref-mZ5zqhmN" role="doc-biblioref">24</a>]</span>;<span class="citation" data-cites="1353itlbv">[<a href="#ref-1353itlbv" role="doc-biblioref">25</a>]</span>;<span class="citation" data-cites="bniVH7q2">[<a href="#ref-bniVH7q2" role="doc-biblioref">35</a>]</span>.
However, most tie themselves tightly to a particular standard format or modelling language ecosystem, thereby inheriting many of the limitations described above.
The Knowledge Graph Hub provides a data loader pipeline, KGX allows conversion of KGs between different technical formats, and RTX-KG2 builds a fixed semantically standardised KG; all three adhere to the Biolink model <span class="citation" data-cites="1353itlbv bniVH7q2">[<a href="#ref-1353itlbv" role="doc-biblioref">25</a>,<a href="#ref-bniVH7q2" role="doc-biblioref">35</a>]</span>.
Bio2BEL is an extensive framework to transform primary databases into BEL <span class="citation" data-cites="mZ5zqhmN">[<a href="#ref-mZ5zqhmN" role="doc-biblioref">24</a>]</span>.
PheKnowLator is the only tool that is conceptually similar to BioCypher in that it allows the creation of knowledge graphs under different data models 14.
However, it appears to be aimed at knowledge representation experts, requiring considerable bioinformatics and ontology expertise.
While being fully customisable, it does not feature flexible recombination of modular components.</p></li>
</ol>
<p>The strategy of subgraph extraction to yield smaller, user-specific KGs has been implemented previously, for instance by CROssBAR (v1), ROBOKOP, and the BioThings Explorer <span class="citation" data-cites="57ZT8woI">[<a href="#ref-57ZT8woI" role="doc-biblioref">10</a>]</span>;<span class="citation" data-cites="D842bsN2">[<a href="#ref-D842bsN2" role="doc-biblioref">36</a>]</span>;<span class="citation" data-cites="lloXDt6h">[<a href="#ref-lloXDt6h" role="doc-biblioref">37</a>]</span>.
However, these rely on single (and thus enormous) harmonised KGs for extracting the subgraphs as opposed to BioCypher’s modular approach <span class="citation" data-cites="1EHBHnMYB">[<a href="#ref-1EHBHnMYB" role="doc-biblioref">38</a>]</span>.
While the “top-down” approach of first building a massive KG and then extracting subgraphs from it is a valid means to arrive at a particular knowledge representation, the effort involved is detrimental to efficiency and democratisation of the process.
A secondary consequence of this large primary effort is that alternative representations of the initial KG will probably not be attempted, hindering flexible knowledge representation.
In contrast, the “bottom-up” approach we follow in BioCypher emphasises modular recombination and flexible representation with small effort overheads.</p>
<p>Ontology mapping has been leveraged for data integration by consortia such as the Monarch Initiative (which is the parent organisation of the MONDO Disease Ontology and the Biolink model, among others) as well as single projects, such as KaBOB <span class="citation" data-cites="ScyBF3gY J06lwBzy">[<a href="#ref-ScyBF3gY" role="doc-biblioref">39</a>,<a href="#ref-J06lwBzy" role="doc-biblioref">40</a>]</span>.
While conceptually related to BioCypher in the use of ontology and biomedical data, these are massive efforts that are not amenable to replication by the average research group.
We aim to close this gap by providing an agile and modular framework that facilitates the reuse of the valuable resources generated by those projects.</p>
<p>There exist alternatives to workflows that involve KGs.
While the premise of our manuscript is that KGs are an important part of sustainable and trustworthy machine learning in the biomedical sciences, “zero domain knowledge” approaches such as UniHPF <span class="citation" data-cites="fO4n5GVt">[<a href="#ref-fO4n5GVt" role="doc-biblioref">41</a>]</span> can do without prior knowledge in their inference process.
Whether methods that forego knowledge representation entirely can be as good or better than methods that use knowledge representation is still a matter of discussion <span class="citation" data-cites="rf9J6yvQ">[<a href="#ref-rf9J6yvQ" role="doc-biblioref">1</a>]</span>;<span class="citation" data-cites="8CoykoAG">[<a href="#ref-8CoykoAG" role="doc-biblioref">42</a>]</span>;<span class="citation" data-cites="QOL3GPr8">[<a href="#ref-QOL3GPr8" role="doc-biblioref">43</a>]</span>;<span class="citation" data-cites="vy5qjeoS">[<a href="#ref-vy5qjeoS" role="doc-biblioref">44</a>]</span>;<span class="citation" data-cites="CiGapf55">[<a href="#ref-CiGapf55" role="doc-biblioref">3</a>]</span>;<span class="citation" data-cites="1FrsWUfLP">[<a href="#ref-1FrsWUfLP" role="doc-biblioref">45</a>]</span>;<span class="citation" data-cites="8keWQlG8">[<a href="#ref-8keWQlG8" role="doc-biblioref">46</a>]</span>.
One aspect that is apparent from modern developments in large language models is that prior knowledge-free models appear to be very data hungry; while billion parameter models are very impressive in their text and image processing capabilities, we do not nearly have enough data in molecular biomedicine to train a GPT-like model, even if we had the funds to train it.
In addition, even in prior knowledge-free deep models, a semantically enriched knowledge graph can still play a role and be useful as an in-process component <span class="citation" data-cites="erUdG5Ur">[<a href="#ref-erUdG5Ur" role="doc-biblioref">12</a>]</span>.
To address these and other performance-related questions, we want to facilitate the creation of benchmarks and standard datasets through the modular nature of our framework.</p>
<h2 id="supplementary-note-5---case-studies">Supplementary Note 5 - Case studies</h2>
<p>In the following sections, we illustrate the usefulness of various design aspects of BioCypher in practical examples.
For most of these case studies, an actual implementation already exists, while some are still drafts or work in progress in early stages.
Practical implementations including public code can be accessed for Modularity, Tumour board, Network expansion, Subgraph extraction, Embedding, and Open Targets.</p>
<h3 id="modularity">Modularity</h3>
<p>There are several resources used by the biomedical community that can be considered essential to a majority of bioinformatics tasks.
A good example is the curation effort on proteins done by the members of the Universal Protein Resource (UniProt) consortium <span class="citation" data-cites="cvbuQhIz">[<a href="#ref-cvbuQhIz" role="doc-biblioref">17</a>]</span>; many secondary resources and tools depend on consistent and comprehensive annotations of the major actors in molecular biology.
As such, there are an enormous number of individual tools and resources that make requests to the public interface of the UniProt service, all of which need to be individually maintained.
We and several of our close collaborators make use of this resource, for instance in OmniPath <span class="citation" data-cites="BDzI9Y5F">[<a href="#ref-BDzI9Y5F" role="doc-biblioref">9</a>]</span>, CKG <span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span>, Bioteque <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>, and the CROssBAR drug discovery and repurposing database <span class="citation" data-cites="57ZT8woI">[<a href="#ref-57ZT8woI" role="doc-biblioref">10</a>]</span>.
We have created an example on how to share a UniProt adapter between resources and how to use BioCypher to combine pre-existing databases based on ontology.</p>
<p>We have written such an adapter for UniProt data, using software infrastructure provided by the OmniPath backend PyPath (for downloading and locally caching the data).
The adapter provides the data as well as convenient access points and an overview of the available property fields using Python Enum classes, offering automatic suggestion and autocomplete functionality.
Using these methods, selecting specific content from the entirety of UniProt data and integrating this content with other resources is greatly facilitated (Figure S1), since the alternative would be, in many cases, to use a manual script to access the UniProt API and rely on manual harmonisation with other datasets.</p>
<p>Similarly, we have added adapters for protein-protein interactions from the popular sources IntAct 7, BioGRID <span class="citation" data-cites="nTlj6VKs">[<a href="#ref-nTlj6VKs" role="doc-biblioref">47</a>]</span>, and STRING <span class="citation" data-cites="1663S4mH8">[<a href="#ref-1663S4mH8" role="doc-biblioref">48</a>]</span>, as well as other resources.
For an up-to-date overview of the BioCypher pipelines and adapters, please visit the <a href="https://github.com/orgs/biocypher/projects/3">Components board</a> and the <a href="https://meta.biocypher.org">meta-graph</a>.
By using the UniProt accession of proteins in the KG and BioCypher functionality, the sources are seamlessly integrated into the final KG despite their differences in original data representation.
As with UniProt data, access to interaction data is facilitated by provision of Enum classes for the various fields in the original data.
The adapters and a script demonstrating their usage are available <a href="https://github.com/HUBioDataLab/CROssBAR-BioCypher-Migration">on GitHub</a>.
The project uses Biolink version 3.2.1.</p>
<p>Figure S1: Modularity of knowledge input.
Individual primary source adapters can be used to build secondary knowledge curations such as OmniPath (compare to Figure 1A).
This shifts maintenance towards the primary source and thus reduces maintenance effort: instead of maintaining each primary resource at the integrated KG level, only one reusable adapter for each resource is necessary.
The primary adapters provide an additional level of flexibility to the user by providing accessible insight into the contents of each primary resource, which can be extensive.
For instance, in the adapter for the UniProt knowledge base, the user can select their favourite species, fields of protein information such as the length or mass of the protein, and relationships to import, such as the host organism or the coding gene of each protein.</p>
<h3 id="tumour-board">Tumour board</h3>
<p>Cancer patients nowadays benefit from a large range of molecular markers that can be used to establish precise prognoses and direct treatment <span class="citation" data-cites="pyC7TREj eyEtjj2p">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>,<a href="#ref-eyEtjj2p" role="doc-biblioref">49</a>]</span>.
In the context of the DECIDER project (www.deciderproject.eu), we are creating a platform to inform the tumour board of actionable molecular phenotypes of high-grade serous ovarian cancer patients.
The current manual workflow for discovering actionable genetic variants consists of multiple complex database queries to different established cancer genetics databases <span class="citation" data-cites="pyC7TREj 1Hew9xuhZ Prcc1PJY">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>,<a href="#ref-1Hew9xuhZ" role="doc-biblioref">50</a>,<a href="#ref-Prcc1PJY" role="doc-biblioref">51</a>]</span>.
The returns from each of the individual queries then need to be curated by human experts (geneticists) in regard to their identity (e.g.
identify duplicate hits from different databases), biological relevance, level of evidence, and actionability.
The heterogeneous nature of results received from different primary database providers makes this a time-consuming task, and a bottleneck for the discovery and comprehensive evaluation of all possible treatment options.</p>
<p>To facilitate the discovery of actionable variants and reduce the manual labour of human experts, we use BioCypher to transform the individual primary resources into an integrated, task-specific KG.
Through mapping of the contents of each primary resource to ontological classes in the build process, we largely remove the need to manually curate and harmonise the individual database results.
This mapping is determined once, at the beginning of the integration process, and results in a BioCypher schema configuration that details the types of entities in the graph (e.g., patients, different types of variants, related treatment options, etc.) and how they are mapped and thus integrated into the underlying ontological framework.
As a second step, datasets that are not yet available from pre-existing BioCypher adapters are adapted in similar fashion to yield data ready to be ingested by BioCypher.
The code for this project can be found at https://github.com/oncodash/oncodashkb.</p>
<p>We make use of the ontology manipulation facilities provided by BioCypher to extend the broad but basic Biolink ontology in certain branches where it is useful to have more granular information about the data that enters the KG.
For example, the exact type of genetic variants are of high importance in the molecular tumour board process, but Biolink only provides a generic “sequence variant” class in its schema.
Therefore, we extended the ontology tree at this node with the very granular corresponding subtree of the Sequence Ontology (SO, <span class="citation" data-cites="xiGrNkR1">[<a href="#ref-xiGrNkR1" role="doc-biblioref">52</a>]</span>), yielding a hybrid ontology with the generality of Biolink and the accuracy of a specialised ontology of sequence variants (Figure S2).
Building on the mechanism provided by BioCypher, this hybridisation can be performed by providing only the minimal input of the sequence ontology URL and the nodes that should be the point of merging (“sequence variant” in Biolink and “sequence_variant” in SO).
The same process is used with the Disease Ontology <span class="citation" data-cites="ez55Zlmu">[<a href="#ref-ez55Zlmu" role="doc-biblioref">53</a>]</span> and OncoTree <span class="citation" data-cites="nDiR3rcY">[<a href="#ref-nDiR3rcY" role="doc-biblioref">54</a>]</span> (see Figure S2).
We use Biolink v3.2.1 and the most recent version of Disease Ontology (as provided by the OBO Foundry at http://purl.obolibrary.org/obo/so.owl).</p>
<p>Figure S2: Modular ontology.
BioCypher combines modular inputs from biomedical resources (left) with a flexible scaffold based on ontology (bottom) to build task-specific knowledge graphs (KGs) with variable format (middle).
Users can configure the use of individual resources as well as the contents taken from these resources in a community-curated collection of adapters.
The data are then harmonised on the basis of user-specified ontologies that are tailored to the specific purpose of the desired KG, using BioCypher’s mapping, extension, and hybridisation facilities.
Finally, the KG is provided to the user through an output adapter in the desired format.
Since Biolink has a broad but general representation of biomedical classes, we extend the “sequence variant” with the corresponding granular information from the specialised Sequence Ontology (right side).
Similarly, information about cancer and specific tumour types are added from Disease Ontology and OncoTree.</p>
<p>Once the database has been created through BioCypher, the process of querying for an actionable variant and its associated treatment options for a given patient is greatly simplified.
This approach also improves the concordance of knowledge base sources, the ability to incorporate external clinical resources, and the recovery of evidence only represented in a single resource <span class="citation" data-cites="pyC7TREj">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>]</span>.</p>
<p>The major advantage of using BioCypher to integrate several resources is the formal representation of the process provided by the schema configuration, which allows for a simple description and long-term centralised maintenance.
Other approaches <span class="citation" data-cites="pyC7TREj">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>]</span> need ad-hoc scripts, hindering refactoring if the input resources change, and lose metadata about the provenance of the merged information, hindering a posteriori analysis.</p>
<h3 id="network-expansion">Network expansion</h3>
<p>Database schemata of large-scale biomedical knowledge providers are tuned for effective storage.
For analysis, the user may benefit from a more dedicated schema type corresponding to the biological question under investigation.
We created BioCypher with the objective to simplify the transformation from storage-optimised schemas to analysis-focused schemas.
Given one or multiple data sources, the user should be able to quickly build a task-specific knowledge graph using only a simple configuration of the desired graph contents.
We demonstrate the simplifying capabilities using an interaction-focussed graph database derived from the Open Targets platform as an example <span class="citation" data-cites="11HFBpGej">[<a href="#ref-11HFBpGej" role="doc-biblioref">55</a>]</span>.</p>
<p>Barrio-Hernandez et al.
used this graph database to inform their method of network expansion <span class="citation" data-cites="9fcv9d5">[<a href="#ref-9fcv9d5" role="doc-biblioref">56</a>]</span>.
The database runs on Neo4j, containing about 9 million nodes and 43 million edges.
It focuses on interactions between biomedical agents such as proteins, DNA/RNA, and small molecules.
Returning one particular interaction from the graph requires a Cypher query of ~13 lines which returns ~15 nodes with ~25 edges (variable depending on the amount of information on each interaction).
A procedure to collect information about these interactions from the graph is provided with the original manuscript <span class="citation" data-cites="9fcv9d5">[<a href="#ref-9fcv9d5" role="doc-biblioref">56</a>]</span>, containing <a href="http://ftp.ebi.ac.uk/pub/databases/intact/various/ot_graphdb/current/apoc_procedures_ot_data.txt">Cypher query code of almost 400 lines</a>.
Still, this extensive query only covers 11 of the 37 source labels, 10 of the 43 target labels, and 24 of the 76 relationship labels that are used in the graph database, offering a large margin for optimisation in creating a task-specific KG.</p>
<p>After BioCypher adaptation, the KG (covering all information used by Barrio-Hernandez et al.) has been reduced to ~700k nodes and 2.6 million edges, a more than ten-fold reduction, without loss of information with regard to this specific task.
This lossless reduction is possible due to 1) the semantic abstraction and 2) the removal of information in the original graph that is not relevant to the task.
Compared to the original file of the database dump (zipped, 1.1 GB), the BioCypher output is ~20-fold smaller (zipped, 63 MB), which greatly facilitates sharing and accessibility (e.g.
by simplifying online access via Jupyter notebooks).
The Cypher query for an interaction has been reduced from 13 query lines, 15 nodes, and 25 edges to 2 query lines, 3 nodes, and 2 edges (Figure S3).
This change comes with a reduction in complexity, which may be beneficial for the experience of interacting with the KG.
If the Cypher query is programmatically generated, this does not play a role for the user.
However, in that case, the complexity is shifted upstream to the code that generates the query.</p>
<p>Figure S3: Semantic abstraction.
A) The original, “storage-oriented” format used by the OTAR KG, displaying one interaction with additional data.
B) The Cypher query to receive one interaction from the OTAR graph.
C) The migrated, “task-oriented” format produced by the BioCypher adapter, displaying one interaction.
The “additional data” from (A) about experiment and evidence type can be added to the interaction node as a property or encoded in additional nodes connected to the interaction node.
D) The Cypher query to receive one interaction from the migrated graph.</p>
<p>Most of this reduction is due to removal of information that is not relevant to the task at hand and semantic abstraction; for instance, the original chain of <code>(“hgnc”)-[:database]-(“SNAI1”)-[:preferredIdentifier]-(:Interactor)-[:interactorB]-(:Interaction)-[:interactorA]-(:Interactor)-[:preferredIdentifier]-(“EP300”)-[:database]-(“hgnc”)</code> to qualify one protein-protein-interaction can be reduced to <code>(“EP300”)-[:enzyme]-(“phosphorylation”)-[:enzyme target]-(“SNAI1”)</code>.
Arguably, the shorter BioCypher query is also more informative, since it details the type of interaction as well as the roles of the participants.
In addition, this representation returns sources of information about the proteins and the interaction as properties on the nodes, and the hierarchical ontology-derived labels provide rich information about the biological context.
For instance, the first ancestor labels of the “phosphorylation” node are “enzymatic interaction”, “direct interaction”, and “physical association”, grounding this specific interaction in its biological context and enabling flexible queries for broader or more specific terms.
This additional information was introduced into the data model by combining the Biolink ontology with the molecular interaction ontology by the Proteomics Standards Initiative <span class="citation" data-cites="KwOAgLE2">[<a href="#ref-KwOAgLE2" role="doc-biblioref">32</a>]</span>.
Thus, this “task-oriented” representation is complementary to the “storage-oriented” one, serving a different purpose, and BioCypher provides an easy and reliable way of going from one type of representation to the other.</p>
<p>The BioCypher migration is fast (about 15 minutes on a common laptop) and tested end-to-end, including deduplication of entities and relationships as well as verbose information on violations of the desired structure (e.g., due to inconsistencies in the input data), making the user explicitly aware of any fault points.
Through this feedback, several inconsistencies were found in the original Open Targets graph during the migration, some of which originated from misannotation in the SIGNOR primary resource (e.g., “P0C6X7_PRO_0000037309” and “P17861_P17861-2”).
This problem affected only a few proteins, which could have gone unnoticed in a manual curation of the data; a type of problem that likely is common in current collections of biomedical knowledge.</p>
<p>Knowledge representations can and should be tuned according to the specific needs of the downstream task to be performed; BioCypher is designed to accommodate arbitrarily simple or complex representations while retaining information important to biomedical research tasks.
A compressed structure is important, for instance, in graph machine learning and embedding tasks, where each additional relationship exponentially increases computational effort for message passing and embedding techniques 12,<span class="citation" data-cites="iDcPtWv1">[<a href="#ref-iDcPtWv1" role="doc-biblioref">57</a>]</span>.
Most importantly, evidence (which experiment and publication the knowledge is derived from) and provenance (who provided which aspects of the primary data) should always be propagated.
The former is essential to enable accurate confidence measures, e.g., not double-counting the same information because it was derived from two secondary sources which refer to the same original publication.
The latter is important for attribution of work that the primary maintainers of large collections of biomedical knowledge provide to the community.
The code of this migration can be found at https://github.com/biocypher/open-targets.
The project uses Biolink v3.2.1.</p>
<h3 id="subgraph-extraction">Subgraph extraction</h3>
<p>For many practical tasks in the workflow of a research scientist, the full KG is not required.
For this reason, building complex and extensive KGs such as the CKG <span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span> or the Bioteque <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span> would not be sensible in all use cases.</p>
<p>For instance, in the context of a proteomics analysis, the user would only like to contextualise their list of differentially abundant proteins using literature connections in the CKG, rendering much of the information on genetics and clinical parameters unnecessary.
In addition, the KG may contain sensitive data on previous projects or patient samples, which cannot be shared (e.g.
in the case of publishing the analysis), causing reproducibility issues.
Likewise, some datasets cannot be shared due to their licences.
With BioCypher, a subset of the entire knowledge collection can be quickly and easily created, taking care to not include sensitive, irrelevant, or unlicensed data.
The analyst merely needs to select the relevant species (e.g.
proteins, diseases, and articles) and their relationships in the BioCypher configuration.
BioCypher then queries the original KG and extracts the required knowledge, conserving all provenance information, and yielding a much-reduced data set ready for sharing.</p>
<p>The original CKG is shared as a Neo4j database dump with a compressed size of 5-7 GB (depending on the version), including 15M nodes and 188M edges.
After BioCypher migration of the full CKG, the same KG can be created from BioCypher output files that have a compressed size of 1.3 GB.
Of note, the creation from BioCypher files using the admin import command is Neo4j version-independent, which is not the case for dump files and can be a reproducibility issue for earlier versions; for instance, the graph of Barrio-Hernandez et al.
in the “Network expansion” case study is a Neo4j v3 dump, which is no longer supported by the current Neo4j Desktop application.
Finally, after the subsetting procedure, the reduced KG (including 5M nodes and 50M edges) in BioCypher format has a compressed size of 333 MB.</p>
<p>Since a complete <a href="https://github.com/biocypher/clinical-knowledge-graph">CKG adapter</a> already existed, the subsetting required minimal effort; i.e., the only required step was to remove unwanted contents from the complete schema configuration.
The code for this task can be found in the same <a href="https://github.com/biocypher/clinical-knowledge-graph/blob/main/scripts/subset_ckg_script.py">repository</a>.
This project uses Biolink v3.2.1.</p>
<h3 id="embedding">Embedding</h3>
<p>As a second subsetting example, we demonstrate the usefulness of subsetting KGs for task-specific graph embeddings.
KG embeddings can be an efficient lower-dimensional substitute for the original data in many machine learning tasks 12 and, as methods such as GEARS <span class="citation" data-cites="Tm14Ibs7">[<a href="#ref-Tm14Ibs7" role="doc-biblioref">58</a>]</span> show, these embeddings can be useful for very complex, hard tasks.
However, including all prior data in every embedding is not necessary for good results, while using the proper domain of knowledge can vastly increase the performance of downstream tasks <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>.
This issue extends both to the type of knowledge represented (not every kind of relationship is relevant to any given task) as well as the source of the knowledge (different focus points in knowledge resources lead to differential performance across different tasks).
Thus, it is highly desirable to have a means to identify the proper knowledge domain relevant to a specific task to increase the efficiency of subsequent analyses.</p>
<p>To achieve this aim, BioCypher can facilitate task-specific builds of well-defined sets of knowledge from a combination of primary sources for each application scenario.
And, since the BioCypher framework automates much of the build process going from only a simple configuration file, the knowledge representations can be iterated over quickly to identify the most pertinent ones.
As above, the only requirement from the user (given existing BioCypher adapters for all requested primary sources) is a selection of biological entities and their relationships in the schema configuration.</p>
<p>We have performed this method of subsetting embedding in the Bioteque environment <span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span> with a subset of the Clinical Knowledge Graph <span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span>.
Concretely, we emulated a scenario where a user seeks to computationally describe the patient samples available in the CKG to explore context-specific similarity between patients.
In brief, we first selected a few sequences of relationships (i.e.
the metapath) to connect subjects (patients) to the proteins expressed by their individual samples, (i.e.
subject → biological sample → analytical sample → protein).
Given the rich variety of associations available for protein entities, we can further link these subjects to other entities and relations available in the knowledge graph, enabling the exploration of specific contexts.
For instance, we extended the metapath to connect the subjects’ protein readouts to biological pathways.
Importantly, due to the gigantic size of the CKG, it was fundamental to use a CKG BioCypher adapter to extract the pertinent subgraphs containing only the required knowledge (e.g., patient-protein data and pathways).
Indeed, selecting the desired KG entities from the complete adapter required negligible time (demonstrated at https://github.com/biocypher/clinical-knowledge-graph).
Finally, the protein- and pathway-based patient descriptors were obtained by running the Bioteque embedding pipeline (https://gitlabsbnb.irbbarcelona.org/bioteque/).
The two resulting patient embedding spaces and their corresponding cluster similarity are provided in Figure S4.</p>
<p>Figure S4.
Bioteque-based patient embeddings.
Two embedding spaces were obtained to describe patients (‘Subjects’ nodes in the CKG) based on protein (left) and protein-pathway (right) similarities.
a, d: 2D projection (t-SNE) of the subjects according to the protein similarity (a) and pathway similarity (d).
Subject nodes are coloured by disease type (see legend at the bottom) while proteins (a) and pathways (d) are coloured in grey.
b, e: Assessment of the quality of the embeddings by quantifying their ability to reassemble the original network.
For each edge in the original network, we compute the cosine distance between its constituent nodes using the node embeddings.
We then generate 100 random permutations for each edge in the network, preserving the degree of each node, and calculate the cosine distances between them.
Finally, for each permutation we sorted all the distances and computed the ROC and PR curves using the original network edges and the corresponding random permutation as the positive and negative sets, respectively.
c, f: Heatmap showing the cosine similarity of the subject embeddings.
To make similarities comparable between heatmaps, cosine similarities were transformed into z-scores by subtracting the median and dividing by the IQR of their corresponding background distribution.
Thus, the higher/redder the z-score the higher the similarity.
An agglomerative hierarchical clustering based on the protein-driven similarities (c) was used to sort the rows and columns in both heatmaps (c, f).
Rows (columns) are coloured based on the subject’s disease (see legend at the bottom).
Notice how, while both spaces are unsurprisingly similar (i.e., both are based on protein readouts), the pathway-driven similarities reveal sub-clusters within disease types that were not evident based on purely protein-driven similarities.</p>
<p>Note that, thanks to the modular nature of the Bioteque pipeline, it is possible to generate embeddings from any network (even beyond the ones used in the Bioteque KG) by just extracting the connections forming the metapath.
In this regard, BioCypher offers a handy means to query the pertinent input files for the Bioteque pipeline, paving the way for an efficient exploration, identification, and extraction of task-specific KG contexts (e.g., generation of KG embeddings for patient similarity exploration).
Indeed, a similar exercise can be performed on the Open Targets dataset (see next section), with minimal preparatory effort.
This would allow, for instance, to further connect protein readouts to disease associations or to complement patient descriptors with embeddings of diseases, drugs, and drug targets for downstream predictive pipelines.</p>
<h3 id="open-targets">Open Targets</h3>
<p>The Open Targets platform is an open resource for drug discovery provided by the European Bioinformatics and Sanger Institutes <span class="citation" data-cites="11HFBpGej">[<a href="#ref-11HFBpGej" role="doc-biblioref">55</a>]</span>.
Their core dataset on drug target-disease relationships is provided for download in columnar format; it is internally harmonised but only partially mapped to several disjoint ontologies (mainly disease-related).
The dataset can be downloaded in Parquet format, a data structure designed to work on distributed systems in a highly parallel manner, making efficient BioCypher adaptation very simple.</p>
<p>To enable an open, community-maintained KG version of the columnar Open Targets dataset, we created a BioCypher adapter using Biolink v3.2.1 (https://github.com/biocypher/open-targets).
Due to the efficient data processing using Parquet and PySpark, the adapter can be run on small machines such as current laptops as well as in distributed high-performance computing environments.
It provides a flexible basis for individually customised KGs from Open Targets data and allows frequent rebuilding of the KGs when the dataset is updated.
The simple layout of a BioCypher adapter allows rapid implementation (less than 500 lines of code) and response to breaking changes in the source material (such as structural or name changes).
Additionally, since the adapter can be reused, changes need to be implemented only once for the benefit of all downstream users.</p>
<p>As shown in the case study “Modularity”, user access of the data is facilitated by Enum classes detailing the dataset contents, allowing automatic suggestions and autocomplete, including all individual source datasets.
Licences of all original data are propagated, and the use of BioCypher “strict mode” guarantees the inclusion of licence, source, and version fields on every single entity of the KG, greatly simplifying downstream decisions related to licensing.</p>
<p>Mapping the Open Targets dataset to a central ontology also facilitates integration with further datasets such as UniProt and the Cancer Dependency Map.
Since Open Targets is a gene-centric platform, data from UniProt can yield complementary insights on the protein layer, for instance by coupling to other datasets of signalling cascades.
We included information on human proteins by simply adding the protein node type and the gene-to-protein edge from the UniProt adapter described in section Modularity.
Harmonising the data was then a simple matter of loading the additional adapter, making sure that the identifier namespace used for genes (ENSEMBL gene) was the same in both adapters (via Enum-based configuration), and writing the information to disk via BioCypher.
It only required the addition of 8 lines of code in the build script.
Adding gene essentiality and cell line information from the Dependency Map project adapter was performed similarly by adding the adapter and loading nodes and edges in the correct format.</p>
<h3 id="federated-learning">Federated learning</h3>
<p>Federated learning is a machine learning approach that enables multiple parties to collaboratively train a shared model while keeping their data decentralised and private <span class="citation" data-cites="11A8WJC5q aDwLmWW9">[<a href="#ref-11A8WJC5q" role="doc-biblioref">59</a>,<a href="#ref-aDwLmWW9" role="doc-biblioref">60</a>]</span>.
This is achieved by allowing each party to train a local version of the model on their own data, and then sharing the updated model parameters with a central server that aggregates these updates.
However, most machine learning algorithms depend on a unified structure of the input; when it comes to algorithms that combine prior knowledge with patient data, a large amount of harmonisation needs to occur before the algorithms can be applied.</p>
<p>BioCypher facilitates federated machine learning by providing an unambiguous blueprint for the process of mapping input data to ontology.
Once a schema for a specific machine learning task has been decided on by the organisers, the BioCypher schema configuration can be distributed, ensuring the same database layout in all training instances.
The usefulness becomes apparent in two pilot projects outlined below.</p>
<p>Firstly, the Care-for-Rare project of the Munich Children’s Hospital has to synchronise a broad spectrum of biomedical data: demographics, medical history, medical diagnosis, laboratory results from routine diagnostics, imaging and omics data with analyses of proteome, metabolome and transcriptome in different tissues as well as genetic information.
To allow reaching a sample size that is suitable for modern methods of diagnosis and treatment options in rare diseases 38, world-wide collaboration between children’s hospitals is a necessity.
The unstructured nature of most clinical data necessitates a harmonisation step with subtle challenges with respect to ontology.
For instance, general classifications such as ICD10-GM subsume rare childrens’ diseases under umbrella terms for whole disease groups, requiring alternative coding catalogues such as Orphanet OrphaCodes <span class="citation" data-cites="19oj9JJD5">[<a href="#ref-19oj9JJD5" role="doc-biblioref">61</a>]</span> and the German Alpha-ID <span class="citation" data-cites="nELyJbhp">[<a href="#ref-nELyJbhp" role="doc-biblioref">62</a>]</span>.
Larger ontologies such as HPO <span class="citation" data-cites="o7nDQFO1">[<a href="#ref-o7nDQFO1" role="doc-biblioref">63</a>]</span> and SNOMED-CT <span class="citation" data-cites="6uGIWm7L">[<a href="#ref-6uGIWm7L" role="doc-biblioref">64</a>]</span> are complex and expanded constantly.
In addition to the technical challenges, the legal requirements of patient confidentiality and data protection necessitate extreme care in the processing of all data, hindering information sharing between collaborators.
All of the above poses great challenges in data integration in the clinical setting.</p>
<p>Secondly, the MeDaX project (bioMedical Data eXploration at University Medicine Greifswald) develops innovative and efficient methods for storage, enrichment, comparison, and retrieval of biomedical data based on KG technology.
Embedded in the Medical Informatics Initiative (MII) Germany, MeDaX builds on the federated storage structure for biomedical health care and research data established in all Data Integration Centres (DICs) at German university hospitals.
We envision extending the existing MIRACOLIX toolbox <span class="citation" data-cites="yKZ8alU5">[<a href="#ref-yKZ8alU5" role="doc-biblioref">65</a>]</span> with the MeDaX pipeline to set up local KGs, combining complex heterogeneous data from multiple resources: in addition to biomedical data available only at the DICs due to patient privacy, we include the MII core data set <span class="citation" data-cites="eFQyj5Wh">[<a href="#ref-eFQyj5Wh" role="doc-biblioref">66</a>]</span>, local population studies <span class="citation" data-cites="lrjMS8FK u1j8EALG">[<a href="#ref-lrjMS8FK" role="doc-biblioref">67</a>,<a href="#ref-u1j8EALG" role="doc-biblioref">68</a>]</span>, biomedical ontologies <span class="citation" data-cites="18dC6FMrj">[<a href="#ref-18dC6FMrj" role="doc-biblioref">69</a>]</span>, and public information portals <span class="citation" data-cites="nC3lHM8C">[<a href="#ref-nC3lHM8C" role="doc-biblioref">70</a>]</span>.
BioCypher’s ontology mapping process facilitates future integration of additional data sources (see also the case study “Data integration”).</p>
<p>We enable federated learning pipelines by supplying build instructions for each local database in the form of the schema configuration that can be publicly and centrally maintained, since it contains no sensitive data (Figure S5).
At each training location, a task-specific KG is created from public data (e.g., with the Clinical Knowledge Graph as baseline), using the subsetting facilities described in the case study “Subgraph extraction”.
Afterwards, the sensitive patient data (e.g., germ-line genetic variants) are integrated into this KG at each location, using the BioCypher schema configuration to specify the type of data involved (e.g., clinical measurements, genetic profiling).
This ensures that, regardless of how the sensitive data are represented at each location, the machine learning algorithm works with the exact same structure of KG, preventing accidental or malicious data leakage in the federated learning step.</p>
<p>Figure S5: User interface.
BioCypher provides high-level users with an abstracted pipeline interface that is used to aggregate data from primary adapters while collecting and unifying the individual data inputs.
Configuration needs to take place only globally when combining adapters that provide overlapping identifier systems, which can be assessed through the pipeline interface.
This is useful to synchronise proprietary or sensitive data between single locations in a federated learning pipeline, since the adapters that contain non-public data only need to provide non-sensitive, summary level information about the data they supply.</p>
<h3 id="data-integration">Data integration</h3>
<p>Biomedical data collections are growing to enormous sizes, which makes the handling of data alone a non-trivial task.
Additionally, these large corpuses then need to be put to good use in downstream analyses, including collaborations between groups or even institutions.
The growth of arbitrarily organised large-scale collections of knowledge poses major challenges to the maintainers of these databases:</p>
<p>Maintaining data ingestion pipelines for dozens of upstream data sources is not feasible in a research context and detracts development time from other tasks.</p>
<p>Using a custom (non-standardised) data model, the effort to integrate new upstream data sources grows with the total number of pre-existing data sources.
Each new data source has to be cross-referenced with all existing data sources and inconsistencies arise because the same piece of information may be represented with different levels of abstraction.</p>
<p>The custom data model also complicates collaboration with external researchers.
Integrating data from different contexts requires the collaborators to adapt to the internal data model.</p>
<p>BioCypher can handle all three challenges.
Firstly, the open architecture and community effort around BioCypher allows maintaining core data ingestion pipelines while reusing data adapters from experts in other fields.
Secondly, the well-described data model by virtue of the ontologies used to build the KG drastically reduces the effort required to integrate new data sources because they need only to be adapted to the core data model, not to all existing data.
Thirdly, the combination of an open architecture and ontology-based data integration facilitates collaborations with external researchers.
We maintain two pilot projects for continuous large-scale data integration in a research context, detailed below.</p>
<ol type="1">
<li>The German Centre for Diabetes Research (DZD, www.dzd-ev.de) has developed a knowledge graph to support data integration for translational research.
The internal KG instance provided the foundation of the open-source CovidGraph project <span class="citation" data-cites="QrLcJJPA">[<a href="#ref-QrLcJJPA" role="doc-biblioref">71</a>]</span> which is now maintained by the HealthECCO community (www.healthecco.org).
At the core of the DZD KG is a data ingestion pipeline for PubMed that transforms publication data into a detailed graph representation, including authors, affiliations, references, and MeSH term annotations.
The PubMed graph contains 350 million nodes and 850 million relationships, as well as data on biological entities (genes, transcripts, proteins), their functional annotations, and biochemical interactions.
This KG is used to link internal research data to public knowledge and to generate new research hypotheses.</li>
</ol>
<p>Re-building the data ingestion and maintenance based on BioCypher reduces the time required to bring new data products to researchers at the DZD because the unified data model and ontology-backed data harmonisation allow the reuse of data analysis modules and user interface components.
Removing obstacles for collaboration on the knowledge graph supports interdisciplinary research on diabetes complications and comorbidities.</p>
<ol start="2" type="1">
<li>At the National Centre for Tumour Diseases (NCT) and the German Cancer Research Centre (DKFZ), we aim to integrate a biomedical knowledge graph with patient data from clinical studies, including multi-omics data, to aid in the stratification of novel biomarkers and the implementation of precision medicine.
To achieve this, we are using the BioCypher framework to create a biomedical KG from curated primary data sources.
The KG will be expanded over time through experimental results as well as clinical annotation and will provide an interface for different roles in the cancer research process.
The maintenance and integration of the biomedical knowledge graph with patient data offers new opportunities for analysis that may enhance the accuracy and effectiveness of precision medicine approaches.</li>
</ol>
<p>Upscaling</p>
<p>As biomedical data become larger, integrated analysis pipelines become more expansive and, thus, expensive.
For numerous projects in systems biomedicine to succeed, a flexible way of maintaining and analysing large sets of knowledge is necessary.
This is done most effectively by separating data storage and analysis (such that each component can be individually scaled), while using distributed computing infrastructure to perform both tasks in close vicinity, such as computing clusters.
We have recently published an open-source software, called Sherlock, to perform this type of data management for biomedical projects <span class="citation" data-cites="16z4LZSdJ">[<a href="#ref-16z4LZSdJ" role="doc-biblioref">72</a>]</span>.
However, this pipeline in some ways depends on manual maintenance, for instance in its <a href="https://github.com/earlham-sherlock/earlham-sherlock.github.io/tree/master/loaders">data transformation from primary resource to internal format</a>.</p>
<p>Using BioCypher, we facilitate the maintenance of Sherlock’s input sources by reusing existing adapters and converting the manual scripts to additional adapters for unrepresented resources.
Combined with the unambiguous BioCypher schema configuration, this will make Sherlock’s input side automatable and greatly decrease maintenance effort, unlocking its full potential in managing complex bioinformatics projects and their resources.
Given a configuration that can be developed locally, a project database can be upscaled to arbitrary numbers of nodes on an in-house or commercial cluster just as the project requires, saving compute time and thereby money.
By virtue of the Sherlock-BioCypher integration, these projects retain the benefits from both frameworks; BioCypher provides reusability, transparency, and ontological grounding, while Sherlock makes data storage and analysis vastly more efficient and economical.</p>
<h3 id="contextualization">Contextualization</h3>
<p>Cells communicate with each other by exchanging molecules to organise cell development, tissue homeostasis, or immune reactions <span class="citation" data-cites="HcpxNvpk">[<a href="#ref-HcpxNvpk" role="doc-biblioref">73</a>]</span>.
Recent computational inference strategies have shown that these interactions can be inferred from single-cell transcriptomics data.
Since then, multiple computational tools have been developed to address this task <span class="citation" data-cites="1A6n4edrt">[<a href="#ref-1A6n4edrt" role="doc-biblioref">74</a>]</span>.
However, most of these tools focus on the inference of cell-cell communication (CCC) mediated by proteins, except one recent tool that uses metabolites <span class="citation" data-cites="1EXXNpykM">[<a href="#ref-1EXXNpykM" role="doc-biblioref">75</a>]</span>.</p>
<p>A primary limitation of metabolite-mediated CCC inference from single-cell transcriptomics data is the necessity to estimate metabolite abundance from transcript levels.
To infer metabolite abundances, current methods employ either flux-balance analysis or enrichment-like approaches <span class="citation" data-cites="1EXXNpykM PSwd2RWP 1FhKLN5iq">[<a href="#ref-1EXXNpykM" role="doc-biblioref">75</a>,<a href="#ref-PSwd2RWP" role="doc-biblioref">76</a>,<a href="#ref-1FhKLN5iq" role="doc-biblioref">77</a>]</span>.
The latter require substantial prior knowledge, usually a set of producing and degrading metabolic enzymes for each metabolite, making information about metabolite-receptor interactions essential for deducing CCC.</p>
<p>Existing prior knowledge resources cover each only a small fraction of metabolites produced by most cells (up to 116 <span class="citation" data-cites="1EXXNpykM">[<a href="#ref-1EXXNpykM" role="doc-biblioref">75</a>]</span>).
Further, they lack information of chemical or biological properties that would allow the analysis to focus on specific diseases or tissues.
Thus, a comprehensive resource that enables contextualization to specific biological questions provides a strategy to increase the accuracy of inference approaches, which are known to be highly prone to false positives <span class="citation" data-cites="aNFFB1Q">[<a href="#ref-aNFFB1Q" role="doc-biblioref">78</a>]</span>.</p>
<p>We have integrated the available knowledge about metabolite-receptor interactions that is dispersed across numerous databases.
Metabolic reactions and their corresponding enzymes can be found in databases such as KEGG <span class="citation" data-cites="VwS8W9dk">[<a href="#ref-VwS8W9dk" role="doc-biblioref">79</a>]</span>, REACTOME <span class="citation" data-cites="Gout7oWo">[<a href="#ref-Gout7oWo" role="doc-biblioref">80</a>]</span>, RHEA <span class="citation" data-cites="v9JUscZ1">[<a href="#ref-v9JUscZ1" role="doc-biblioref">81</a>]</span>, HMDB <span class="citation" data-cites="Rygf3Cgf">[<a href="#ref-Rygf3Cgf" role="doc-biblioref">82</a>]</span>, and genome-scale metabolic models such as Recon3D <span class="citation" data-cites="TYquyrUR">[<a href="#ref-TYquyrUR" role="doc-biblioref">83</a>]</span> and Human HMR <span class="citation" data-cites="QANpB4kr">[<a href="#ref-QANpB4kr" role="doc-biblioref">84</a>]</span>.
Meanwhile, information about metabolites and their receptors is available in the STITCH database <span class="citation" data-cites="bc1DWCco">[<a href="#ref-bc1DWCco" role="doc-biblioref">85</a>]</span>, Guide to Pharmacology <span class="citation" data-cites="1GjHUK5qj">[<a href="#ref-1GjHUK5qj" role="doc-biblioref">86</a>]</span>, and Interactomics screens <span class="citation" data-cites="GBpX8fvk">[<a href="#ref-GBpX8fvk" role="doc-biblioref">87</a>]</span>.
All these databases use different identifiers for their metabolite, proteins or reactions, that are often conflicting or redundant <span class="citation" data-cites="171uilzJ7 117E2kGkQ">[<a href="#ref-171uilzJ7" role="doc-biblioref">88</a>,<a href="#ref-117E2kGkQ" role="doc-biblioref">89</a>]</span>.
Using BioCypher, we systematically and reproducibly integrate the knowledge from these databases, facilitating the creation and maintenance of a comprehensive metabolite-receptor interaction database (https://github.com/biocypher/metalinks).</p>
<p>The effectiveness of this approach is exemplified by examining metabolite-mediated CCC in the kidney.
By employing a <a href="https://github.com/biocypher/metalinks/blob/main/cypher_query.txt">few concise lines of Cypher</a>, metabolites and proteins can be filtered to focus on those active in the kidney or present in urine.
Likewise, metabolite-receptor interactions are filtered using confidence levels.
Applying these contextualization parameters reduces the overall size of the dataset by decreasing the number of metabolites from approximately 1400 to a more manageable 394 (derived from enzyme sets), and metabolite-receptor interactions from ~ 100 000 to 3864, featuring 807 unique receptors and 261 unique metabolites.
The resulting table can either be used in Python directly via BioCypher’s support of Pandas data frames, or exported to CSV from Neo4j, and seamlessly integrated into downstream analysis tools performing CCC, such as LIANA <span class="citation" data-cites="aNFFB1Q">[<a href="#ref-aNFFB1Q" role="doc-biblioref">78</a>]</span>.</p>
<p>Supplementary table 1. Recent biomedical knowledge graph solutions (non-comprehensive).
Database
Reference
Biological Insight Knowledge Graph
<span class="citation" data-cites="SmbvHVqc">[<a href="#ref-SmbvHVqc" role="doc-biblioref">22</a>]</span>
Bioteque
<span class="citation" data-cites="vhsYX2y">[<a href="#ref-vhsYX2y" role="doc-biblioref">2</a>]</span>
Clinical Knowledge Graph
<span class="citation" data-cites="JxZlpZ1">[<a href="#ref-JxZlpZ1" role="doc-biblioref">11</a>]</span>
CROssBAR
<span class="citation" data-cites="57ZT8woI">[<a href="#ref-57ZT8woI" role="doc-biblioref">10</a>]</span>
Dependency Map
<span class="citation" data-cites="RnnKuIPr">[<a href="#ref-RnnKuIPr" role="doc-biblioref">90</a>]</span>
GenomicKB
<span class="citation" data-cites="1GDvfevZN">[<a href="#ref-1GDvfevZN" role="doc-biblioref">91</a>]</span>
HealthECCO Covidgraph
<span class="citation" data-cites="QrLcJJPA">[<a href="#ref-QrLcJJPA" role="doc-biblioref">71</a>]</span>
INDRA CogEx
https://github.com/bgyori/indra_cogex
KG-COVID-19
<span class="citation" data-cites="1353itlbv">[<a href="#ref-1353itlbv" role="doc-biblioref">25</a>]</span>
NIH Funding knowledge graph
<span class="citation" data-cites="BvOaZVwT">[<a href="#ref-BvOaZVwT" role="doc-biblioref">92</a>]</span>
OmniPath
<span class="citation" data-cites="BDzI9Y5F">[<a href="#ref-BDzI9Y5F" role="doc-biblioref">9</a>]</span>
Open Targets
<span class="citation" data-cites="11HFBpGej">[<a href="#ref-11HFBpGej" role="doc-biblioref">55</a>]</span>
PheKnowLator
<span class="citation" data-cites="1H2nqqKV7">[<a href="#ref-1H2nqqKV7" role="doc-biblioref">23</a>]</span>
PORI (Platform for Oncogenic Reporting and Interpretation)
<span class="citation" data-cites="pyC7TREj">[<a href="#ref-pyC7TREj" role="doc-biblioref">26</a>]</span>
PrimeKG
<span class="citation" data-cites="8CoykoAG">[<a href="#ref-8CoykoAG" role="doc-biblioref">42</a>]</span>
RTX-KG2
<span class="citation" data-cites="bniVH7q2">[<a href="#ref-bniVH7q2" role="doc-biblioref">35</a>]</span>
TypeDB
https://github.com/typedb-osi/typedb-bio</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-rf9J6yvQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline"><strong>Graph representation learning in biomedicine and healthcare</strong> <div class="csl-block">Michelle M Li, Kexin Huang, Marinka Zitnik</div> <em>Nature Biomedical Engineering</em> (2022-10-31) <a href="https://doi.org/gq533s">https://doi.org/gq533s</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41551-022-00942-x">10.1038/s41551-022-00942-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36316368">36316368</a></div></div>
</div>
<div id="ref-vhsYX2y" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline"><strong>Integrating and formatting biomedical data as pre-calculated knowledge graph embeddings in the Bioteque</strong> <div class="csl-block">Adrià Fernández-Torras, Miquel Duran-Frigola, Martino Bertoni, Martina Locatelli, Patrick Aloy</div> <em>Nature Communications</em> (2022-09-09) <a href="https://doi.org/gqtdvj">https://doi.org/gqtdvj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-022-33026-0">10.1038/s41467-022-33026-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36085310">36085310</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9463154">PMC9463154</a></div></div>
</div>
<div id="ref-CiGapf55" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline"><strong>Knowledge graphs as tools for explainable machine learning: A survey</strong> <div class="csl-block">Ilaria Tiddi, Stefan Schlobach</div> <em>Artificial Intelligence</em> (2022-01) <a href="https://doi.org/gnstxz">https://doi.org/gnstxz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.artint.2021.103627">10.1016/j.artint.2021.103627</a></div></div>
</div>
<div id="ref-yqoUY2f9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline"><strong>A review of biomedical datasets relating to drug discovery: a knowledge graph perspective</strong> <div class="csl-block">Stephen Bonner, Ian P Barrett, Cheng Ye, Rowan Swiers, Ola Engkvist, Andreas Bender, Charles Tapley Hoyt, William L Hamilton</div> <em>Briefings in Bioinformatics</em> (2022-09-23) <a href="https://doi.org/gqv3s3">https://doi.org/gqv3s3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bib/bbac404">10.1093/bib/bbac404</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36151740">36151740</a></div></div>
</div>
<div id="ref-TNqtTxS8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline"><strong>Knowledge-Based Biomedical Data Science</strong> <div class="csl-block">Tiffany J Callahan, Ignacio J Tripodi, Harrison Pielke-Lombardo, Lawrence E Hunter</div> <em>Annual Review of Biomedical Data Science</em> (2020-07-20) <a href="https://doi.org/ghtkzt">https://doi.org/ghtkzt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1146/annurev-biodatasci-010820-091627">10.1146/annurev-biodatasci-010820-091627</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33954284">33954284</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8095730">PMC8095730</a></div></div>
</div>
<div id="ref-6DjakjNS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline"><strong>The FAIR Guiding Principles for scientific data management and stewardship</strong> <div class="csl-block">Mark D Wilkinson, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, Jan-Willem Boiten, Luiz Bonino da Silva Santos, Philip E Bourne, … Barend Mons</div> <em>Scientific Data</em> (2016-03-15) <a href="https://doi.org/bdd4">https://doi.org/bdd4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/sdata.2016.18">10.1038/sdata.2016.18</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26978244">26978244</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4792175">PMC4792175</a></div></div>
</div>
<div id="ref-MMp6cQ1x" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline"><strong>The TRUST Principles for digital repositories</strong> <div class="csl-block">Dawei Lin, Jonathan Crabtree, Ingrid Dillo, Robert R Downs, Rorie Edmunds, David Giaretta, Marisa De Giusti, Hervé L’Hours, Wim Hugo, Reyna Jenkyns, … John Westbrook</div> <em>Scientific Data</em> (2020-05-14) <a href="https://doi.org/ggwrtj">https://doi.org/ggwrtj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41597-020-0486-7">10.1038/s41597-020-0486-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32409645">32409645</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7224370">PMC7224370</a></div></div>
</div>
<div id="ref-1ALlMCWlM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><strong>The Biomedical Resource Ontology (BRO) to enable resource discovery in clinical and translational research</strong> <div class="csl-block">Jessica D Tenenbaum, Patricia L Whetzel, Kent Anderson, Charles D Borromeo, Ivo D Dinov, Davera Gabriel, Beth Kirschner, Barbara Mirel, Tim Morris, Natasha Noy, … Peter Lyster</div> <em>Journal of Biomedical Informatics</em> (2011-02) <a href="https://doi.org/c66d8c">https://doi.org/c66d8c</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.jbi.2010.10.003">10.1016/j.jbi.2010.10.003</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20955817">20955817</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3050430">PMC3050430</a></div></div>
</div>
<div id="ref-BDzI9Y5F" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><strong>OmniPath: guidelines and gateway for literature-curated signaling pathway resources</strong> <div class="csl-block">Dénes Türei, Tamás Korcsmáros, Julio Saez-Rodriguez</div> <em>Nature Methods</em> (2016-11-29) <a href="https://doi.org/gh29s5">https://doi.org/gh29s5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nmeth.4077">10.1038/nmeth.4077</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27898060">27898060</a></div></div>
</div>
<div id="ref-57ZT8woI" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline"><strong>CROssBAR: comprehensive resource of biomedical relations with knowledge graph representations</strong> <div class="csl-block">Tunca Doğan, Heval Atas, Vishal Joshi, Ahmet Atakan, Ahmet Sureyya Rifaioglu, Esra Nalbat, Andrew Nightingale, Rabie Saidi, Vladimir Volynkin, Hermann Zellner, … Volkan Atalay</div> <em>Nucleic Acids Research</em> (2021-06-28) <a href="https://doi.org/gndrgn">https://doi.org/gndrgn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab543">10.1093/nar/gkab543</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34181736">34181736</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8450100">PMC8450100</a></div></div>
</div>
<div id="ref-JxZlpZ1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline"><strong>A knowledge graph to interpret clinical proteomics data</strong> <div class="csl-block">Alberto Santos, Ana R Colaço, Annelaura B Nielsen, Lili Niu, Maximilian Strauss, Philipp E Geyer, Fabian Coscia, Nicolai JWewer Albrechtsen, Filip Mundt, Lars Juhl Jensen, Matthias Mann</div> <em>Nature Biotechnology</em> (2022-01-31) <a href="https://doi.org/gpbx68">https://doi.org/gpbx68</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41587-021-01145-6">10.1038/s41587-021-01145-6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/35102292">35102292</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9110295">PMC9110295</a></div></div>
</div>
<div id="ref-erUdG5Ur" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><strong>Enhanced Story Comprehension for Large Language Models through Dynamic Document-Based Knowledge Graphs</strong> <div class="csl-block">Berkeley R Andrus, Yeganeh Nasiri, Shilong Cui, Benjamin Cullen, Nancy Fulda</div> <em>Proceedings of the AAAI Conference on Artificial Intelligence</em> (2022-06-28) <a href="https://doi.org/gs5th3">https://doi.org/gs5th3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1609/aaai.v36i10.21286">10.1609/aaai.v36i10.21286</a></div></div>
</div>
<div id="ref-11YmMTmFW" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><strong>A Platform for the Biomedical Application of Large Language Models</strong> <div class="csl-block">Sebastian Lobentanzer, Julio Saez-Rodriguez</div> <em>arXiv</em> (2023) <a href="https://doi.org/gs5th4">https://doi.org/gs5th4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2305.06488">10.48550/arxiv.2305.06488</a></div></div>
</div>
<div id="ref-vzPcrJo9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline"><strong>Biolink Model: A universal schema for knowledge graphs in clinical, biomedical, and translational science</strong> <div class="csl-block">Deepak R Unni, Sierra AT Moxon, Michael Bada, Matthew Brush, Richard Bruskiewich, JHarry Caufield, Paul A Clemons, Vlado Dancik, Michel Dumontier, Karamarie Fecho, … </div> <em>Clinical and Translational Science</em> (2022-06-06) <a href="https://doi.org/gq6hdn">https://doi.org/gq6hdn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/cts.13302">10.1111/cts.13302</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36125173">36125173</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9372416">PMC9372416</a></div></div>
</div>
<div id="ref-CNP39tah" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">15. </div><div class="csl-right-inline"><strong>The OBO Foundry: coordinated evolution of ontologies to support biomedical data integration</strong> <div class="csl-block">Barry Smith, Michael Ashburner, Cornelius Rosse, Jonathan Bard, William Bug, Werner Ceusters, Louis J Goldberg, Karen Eilbeck, Amelia Ireland, … Suzanna Lewis</div> <em>Nature Biotechnology</em> (2007-11) <a href="https://doi.org/bqng99">https://doi.org/bqng99</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nbt1346">10.1038/nbt1346</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17989687">17989687</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2814061">PMC2814061</a></div></div>
</div>
<div id="ref-fKdMnXba" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">16. </div><div class="csl-right-inline"><strong>Unifying the identification of biomedical entities with the Bioregistry</strong> <div class="csl-block">Charles Tapley Hoyt, Meghan Balk, Tiffany J Callahan, Daniel Domingo-Fernández, Melissa A Haendel, Harshad B Hegde, Daniel S Himmelstein, Klas Karis, John Kunze, Tiago Lubiana, … Benjamin M Gyori</div> <em>Scientific Data</em> (2022-11-19) <a href="https://doi.org/gq92zt">https://doi.org/gq92zt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41597-022-01807-3">10.1038/s41597-022-01807-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36402838">36402838</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9675740">PMC9675740</a></div></div>
</div>
<div id="ref-cvbuQhIz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">17. </div><div class="csl-right-inline"><strong>UniProt: a hub for protein information</strong> <em>Nucleic Acids Research</em> (2014-10-27) <a href="https://doi.org/f64xfr">https://doi.org/f64xfr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gku989">10.1093/nar/gku989</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25348405">25348405</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4384041">PMC4384041</a></div></div>
</div>
<div id="ref-ezOoICHE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><strong>COSMIC: exploring the world's knowledge of somatic mutations in human cancer</strong> <div class="csl-block">Simon A Forbes, David Beare, Prasad Gunasekaran, Kenric Leung, Nidhi Bindal, Harry Boutselakis, Minjie Ding, Sally Bamford, Charlotte Cole, Sari Ward, … Peter J Campbell</div> <em>Nucleic Acids Research</em> (2014-10-29) <a href="https://doi.org/f64ng8">https://doi.org/f64ng8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gku1075">10.1093/nar/gku1075</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25355519">25355519</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4383913">PMC4383913</a></div></div>
</div>
<div id="ref-dGRWxzm4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><strong>IntAct: an open source molecular interaction database</strong> <div class="csl-block">H Hermjakob</div> <em>Nucleic Acids Research</em> (2004-01-01) <a href="https://doi.org/dz63qs">https://doi.org/dz63qs</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkh052">10.1093/nar/gkh052</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/14681455">14681455</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC308786">PMC308786</a></div></div>
</div>
<div id="ref-NQaiGeFN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">20. </div><div class="csl-right-inline"><strong>Issues in the Registration of Clinical Trials</strong> <div class="csl-block">Deborah A Zarin, Nicholas C Ide, Tony Tse, William R Harlan, Joyce C West, Donald AB Lindberg</div> <em>JAMA</em> (2007-05-16) <a href="https://doi.org/bqg9j8">https://doi.org/bqg9j8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1001/jama.297.19.2112">10.1001/jama.297.19.2112</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17507347">17507347</a></div></div>
</div>
<div id="ref-1GEdqX9gO" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">21. </div><div class="csl-right-inline"><strong>Integrative Transcriptomics Reveals Sexually Dimorphic Control of the Cholinergic/Neurokine Interface in Schizophrenia and Bipolar Disorder</strong> <div class="csl-block">Sebastian Lobentanzer, Geula Hanin, Jochen Klein, Hermona Soreq</div> <em>Cell Reports</em> (2019-10) <a href="https://doi.org/ghm25b">https://doi.org/ghm25b</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.celrep.2019.09.017">10.1016/j.celrep.2019.09.017</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31618642">31618642</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6899527">PMC6899527</a></div></div>
</div>
<div id="ref-SmbvHVqc" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><strong>Biological Insights Knowledge Graph: an integrated knowledge graph to support drug development</strong> <div class="csl-block">David Geleta, Andriy Nikolov, Gavin Edwards, Anna Gogleva, Richard Jackson, Erik Jansson, Andrej Lamov, Sebastian Nilsson, Marina Pettersson, Vladimir Poroshin, … Eliseo Papa</div> <em>Cold Spring Harbor Laboratory</em> (2021-11-01) <a href="https://doi.org/gs52ms">https://doi.org/gs52ms</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.10.28.466262">10.1101/2021.10.28.466262</a></div></div>
</div>
<div id="ref-1H2nqqKV7" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">23. </div><div class="csl-right-inline"><strong>A Framework for Automated Construction of Heterogeneous Large-Scale Biomedical Knowledge Graphs</strong> <div class="csl-block">Tiffany J Callahan, Ignacio J Tripodi, Lawrence E Hunter, William A Baumgartner Jr.</div> <em>Cold Spring Harbor Laboratory</em> (2020-05-02) <a href="https://doi.org/gg338z">https://doi.org/gg338z</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2020.04.30.071407">10.1101/2020.04.30.071407</a></div></div>
</div>
<div id="ref-mZ5zqhmN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">24. </div><div class="csl-right-inline"><strong>Integration of Structured Biological Data Sources using Biological Expression Language</strong> <div class="csl-block">Charles Tapley Hoyt, Daniel Domingo-Fernández, Sarah Mubeen, Josep Marin Llaó, Andrej Konotopez, Christian Ebeling, Colin Birkenbihl, Özlem Muslu, Bradley English, Simon Müller, … Martin Hofmann-Apitius</div> <em>Cold Spring Harbor Laboratory</em> (2019-05-08) <a href="https://doi.org/gg3kpq">https://doi.org/gg3kpq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/631812">10.1101/631812</a></div></div>
</div>
<div id="ref-1353itlbv" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><strong>KG-COVID-19: a framework to produce customized knowledge graphs for COVID-19 response</strong> <div class="csl-block">Justin Reese, Deepak Unni, Tiffany J Callahan, Luca Cappelletti, Vida Ravanmehr, Seth Carbon, Tommaso Fontana, Hannah Blau, Nicolas Matentzoglu, Nomi L Harris, … Christopher J Mungall</div> <em>Cold Spring Harbor Laboratory</em> (2020-08-18) <a href="https://doi.org/ghrpwg">https://doi.org/ghrpwg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2020.08.17.254839">10.1101/2020.08.17.254839</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32839776">32839776</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7444288">PMC7444288</a></div></div>
</div>
<div id="ref-pyC7TREj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">26. </div><div class="csl-right-inline"><strong>A platform for oncogenomic reporting and interpretation</strong> <div class="csl-block">Caralyn Reisle, Laura M Williamson, Erin Pleasance, Anna Davies, Brayden Pellegrini, Dustin W Bleile, Karen L Mungall, Eric Chuah, Martin R Jones, Yussanne Ma, … Steven JM Jones</div> <em>Nature Communications</em> (2022-02-09) <a href="https://doi.org/gp5zvn">https://doi.org/gp5zvn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-022-28348-y">10.1038/s41467-022-28348-y</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/35140225">35140225</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8828759">PMC8828759</a></div></div>
</div>
<div id="ref-DeIC0Dgw" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">27. </div><div class="csl-right-inline"><strong>KGML-xDTD: A Knowledge Graph-based Machine Learning Framework for Drug Treatment Prediction and Mechanism Description</strong> <div class="csl-block">Chunyu Ma, Zhihan Zhou, Han Liu, David Koslicki</div> <em>Cold Spring Harbor Laboratory</em> (2022-12-02) <a href="https://doi.org/gs2hwp">https://doi.org/gs2hwp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.11.29.518441">10.1101/2022.11.29.518441</a></div></div>
</div>
<div id="ref-11W3tHiHT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">28. </div><div class="csl-right-inline"><strong>Recent advances in modeling languages for pathway maps and computable biological networks</strong> <div class="csl-block">Ted Slater</div> <em>Drug Discovery Today</em> (2014-02) <a href="https://doi.org/f5wsb7">https://doi.org/f5wsb7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.drudis.2013.12.011">10.1016/j.drudis.2013.12.011</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24444544">24444544</a></div></div>
</div>
<div id="ref-akdxNohe" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">29. </div><div class="csl-right-inline"><strong>Gene Ontology Causal Activity Modeling (GO-CAM) moves beyond GO annotations to structured descriptions of biological functions and systems</strong> <div class="csl-block">Paul D Thomas, David P Hill, Huaiyu Mi, David Osumi-Sutherland, Kimberly Van Auken, Seth Carbon, James P Balhoff, Laurent-Philippe Albou, Benjamin Good, Pascale Gaudet, … Christopher J Mungall</div> <em>Nature Genetics</em> (2019-09-23) <a href="https://doi.org/ggcfst">https://doi.org/ggcfst</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41588-019-0500-1">10.1038/s41588-019-0500-1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31548717">31548717</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7012280">PMC7012280</a></div></div>
</div>
<div id="ref-mP9wIAFk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">30. </div><div class="csl-right-inline"><strong>&lt;scp&gt;SBML&lt;/scp&gt;
Level 3: an extensible format for the exchange and reuse of biological models</strong> <div class="csl-block">Sarah M Keating, Dagmar Waltemath, Matthias König, Fengkai Zhang, Andreas Dräger, Claudine Chaouiya, Frank T Bergmann, Andrew Finney, Colin S Gillespie, Tomáš Helikar, … </div> <em>Molecular Systems Biology</em> (2020-08) <a href="https://doi.org/gncdt5">https://doi.org/gncdt5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.15252/msb.20199110">10.15252/msb.20199110</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32845085">32845085</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8411907">PMC8411907</a></div></div>
</div>
<div id="ref-1Cn9nrdiq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">31. </div><div class="csl-right-inline"><strong>The BioPAX community standard for pathway data sharing</strong> <div class="csl-block">Emek Demir, Michael P Cary, Suzanne Paley, Ken Fukuda, Christian Lemer, Imre Vastrik, Guanming Wu, Peter D'Eustachio, Carl Schaefer, Joanne Luciano, … Gary D Bader</div> <em>Nature Biotechnology</em> (2010-09) <a href="https://doi.org/fgcrtt">https://doi.org/fgcrtt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nbt.1666">10.1038/nbt.1666</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20829833">20829833</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3001121">PMC3001121</a></div></div>
</div>
<div id="ref-KwOAgLE2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">32. </div><div class="csl-right-inline"><strong>The HUPO PSI's Molecular Interaction format—a community standard for the representation of protein interaction data</strong> <div class="csl-block">Henning Hermjakob, Luisa Montecchi-Palazzi, Gary Bader, Jérôme Wojcik, Lukasz Salwinski, Arnaud Ceol, Susan Moore, Sandra Orchard, Ugis Sarkans, Christian von Mering, … Rolf Apweiler</div> <em>Nature Biotechnology</em> (2004-01-30) <a href="https://doi.org/fsdgcx">https://doi.org/fsdgcx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nbt926">10.1038/nbt926</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/14755292">14755292</a></div></div>
</div>
<div id="ref-Nxf46J67" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">33. </div><div class="csl-right-inline"><strong>Representations of molecular pathways: an evaluation of SBML, PSI MI and BioPAX</strong> <div class="csl-block">Lena Strömbäck, Patrick Lambrix</div> <em>Bioinformatics</em> (2005-10-18) <a href="https://doi.org/dm6h69">https://doi.org/dm6h69</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/bti718">10.1093/bioinformatics/bti718</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/16234320">16234320</a></div></div>
</div>
<div id="ref-PqRxavyS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">34. </div><div class="csl-right-inline"><strong>Promoting Coordinated Development of Community-Based Information Standards for Modeling in Biology: The COMBINE Initiative</strong> <div class="csl-block">Michael Hucka, David P Nickerson, Gary D Bader, Frank T Bergmann, Jonathan Cooper, Emek Demir, Alan Garny, Martin Golebiewski, Chris J Myers, Falk Schreiber, … Nicolas Le NovÃ¨re</div> <em>Frontiers in Bioengineering and Biotechnology</em> (2015-02-24) <a href="https://doi.org/gf5hmg">https://doi.org/gf5hmg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3389/fbioe.2015.00019">10.3389/fbioe.2015.00019</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25759811">25759811</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4338824">PMC4338824</a></div></div>
</div>
<div id="ref-bniVH7q2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">35. </div><div class="csl-right-inline"><strong>RTX-KG2: a system for building a semantically standardized knowledge graph for translational biomedicine</strong> <div class="csl-block">EC Wood, Amy K Glen, Lindsey G Kvarfordt, Finn Womack, Liliana Acevedo, Timothy S Yoon, Chunyu Ma, Veronica Flores, Meghamala Sinha, Yodsawalai Chodpathumwan, … Stephen A Ramsey</div> <em>BMC Bioinformatics</em> (2022-09-29) <a href="https://doi.org/gqxmkb">https://doi.org/gqxmkb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-022-04932-3">10.1186/s12859-022-04932-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36175836">36175836</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9520835">PMC9520835</a></div></div>
</div>
<div id="ref-D842bsN2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">36. </div><div class="csl-right-inline"><strong>ROBOKOP: an abstraction layer and user interface for knowledge graphs to support question answering</strong> <div class="csl-block">Kenneth Morton, Patrick Wang, Chris Bizon, Steven Cox, James Balhoff, Yaphet Kebede, Karamarie Fecho, Alexander Tropsha</div> <em>Bioinformatics</em> (2019-08-13) <a href="https://doi.org/gj7chw">https://doi.org/gj7chw</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btz604">10.1093/bioinformatics/btz604</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31410449">31410449</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6954664">PMC6954664</a></div></div>
</div>
<div id="ref-lloXDt6h" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">37. </div><div class="csl-right-inline"><strong>Cross-linking BioThings APIs through JSON-LD to facilitate knowledge exploration</strong> <div class="csl-block">Jiwen Xin, Cyrus Afrasiabi, Sebastien Lelong, Julee Adesara, Ginger Tsueng, Andrew I Su, Chunlei Wu</div> <em>BMC Bioinformatics</em> (2018-02-01) <a href="https://doi.org/grfnc4">https://doi.org/grfnc4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-018-2041-5">10.1186/s12859-018-2041-5</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29390967">29390967</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5796402">PMC5796402</a></div></div>
</div>
<div id="ref-1EHBHnMYB" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">38. </div><div class="csl-right-inline"><strong>The Biomedical Data Translator Program: Conception, Culture, and Community</strong> <em>Clinical and Translational Science</em> (2018-11-09) <a href="https://doi.org/gktj4p">https://doi.org/gktj4p</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/cts.12592">10.1111/cts.12592</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30412340">30412340</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6440573">PMC6440573</a></div></div>
</div>
<div id="ref-ScyBF3gY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">39. </div><div class="csl-right-inline"><strong>Navigating the Phenotype Frontier: The Monarch Initiative</strong> <div class="csl-block">Julie A McMurry, Sebastian Köhler, Nicole L Washington, James P Balhoff, Charles Borromeo, Matthew Brush, Seth Carbon, Tom Conlin, Nathan Dunn, Mark Engelstad, … Melissa A Haendel</div> <em>Genetics</em> (2016-08-01) <a href="https://doi.org/f82g27">https://doi.org/f82g27</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1534/genetics.116.188870">10.1534/genetics.116.188870</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27516611">27516611</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4981258">PMC4981258</a></div></div>
</div>
<div id="ref-J06lwBzy" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">40. </div><div class="csl-right-inline"><strong>KaBOB: ontology-based semantic integration of biomedical databases</strong> <div class="csl-block">Kevin M Livingston, Michael Bada, William A Baumgartner Jr, Lawrence E Hunter</div> <em>BMC Bioinformatics</em> (2015-04-23) <a href="https://doi.org/f7kdb3">https://doi.org/f7kdb3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12859-015-0559-3">10.1186/s12859-015-0559-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25903923">25903923</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4448321">PMC4448321</a></div></div>
</div>
<div id="ref-fO4n5GVt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">41. </div><div class="csl-right-inline"><strong>UniHPF : Universal Healthcare Predictive Framework with Zero Domain Knowledge</strong> <div class="csl-block">Kyunghoon Hur, Jungwoo Oh, Junu Kim, Jiyoun Kim, Min Jae Lee, Eunbyeol Cho, Seong-Eun Moon, Young-Hak Kim, Edward Choi</div> <em>arXiv</em> (2022) <a href="https://doi.org/gs52nd">https://doi.org/gs52nd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2211.08082">10.48550/arxiv.2211.08082</a></div></div>
</div>
<div id="ref-8CoykoAG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">42. </div><div class="csl-right-inline"><strong>Building a knowledge graph to enable precision medicine</strong> <div class="csl-block">Payal Chandak, Kexin Huang, Marinka Zitnik</div> <em>Cold Spring Harbor Laboratory</em> (2022-05-01) <a href="https://doi.org/gsx5jg">https://doi.org/gsx5jg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.05.01.489928">10.1101/2022.05.01.489928</a></div></div>
</div>
<div id="ref-QOL3GPr8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">43. </div><div class="csl-right-inline"><strong>Deep learning for diagnosing patients with rare genetic diseases</strong> <div class="csl-block">Emily Alsentzer, Michelle M Li, Shilpa N Kobren, Isaac S Kohane, Marinka Zitnik</div> <em>Cold Spring Harbor Laboratory</em> (2022-12-13) <a href="https://doi.org/gs52mt">https://doi.org/gs52mt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.12.07.22283238">10.1101/2022.12.07.22283238</a></div></div>
</div>
<div id="ref-vy5qjeoS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">44. </div><div class="csl-right-inline"><strong>Deep Bidirectional Language-Knowledge Graph Pretraining</strong> <div class="csl-block">Michihiro Yasunaga, Antoine Bosselut, Hongyu Ren, Xikun Zhang, Christopher D Manning, Percy Liang, Jure Leskovec</div> <em>arXiv</em> (2022) <a href="https://doi.org/gs52nb">https://doi.org/gs52nb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2210.09338">10.48550/arxiv.2210.09338</a></div></div>
</div>
<div id="ref-1FrsWUfLP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">45. </div><div class="csl-right-inline"><strong>Knowledge Graph - Deep Learning: A Case Study in Question Answering in Aviation Safety Domain</strong> <div class="csl-block">Ankush Agarwal, Raj Gite, Shreya Laddha, Pushpak Bhattacharyya, Satyanarayan Kar, Asif Ekbal, Prabhjit Thind, Rajesh Zele, Ravi Shankar</div> <em>arXiv</em> (2022) <a href="https://doi.org/gs52m9">https://doi.org/gs52m9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2205.15952">10.48550/arxiv.2205.15952</a></div></div>
</div>
<div id="ref-8keWQlG8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">46. </div><div class="csl-right-inline"><strong>STonKGs: a sophisticated transformer trained on biomedical text and knowledge graphs</strong> <div class="csl-block">Helena Balabin, Charles Tapley Hoyt, Colin Birkenbihl, Benjamin M Gyori, John Bachman, Alpha Tom Kodamullil, Paul G Plöger, Martin Hofmann-Apitius, Daniel Domingo-Fernández</div> <em>Bioinformatics</em> (2022-01-05) <a href="https://doi.org/gqc94q">https://doi.org/gqc94q</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btac001">10.1093/bioinformatics/btac001</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34986221">34986221</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8896635">PMC8896635</a></div></div>
</div>
<div id="ref-nTlj6VKs" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">47. </div><div class="csl-right-inline"><strong>The &lt;scp&gt;BioGRID&lt;/scp&gt; database: A comprehensive biomedical resource of curated protein, genetic, and chemical interactions</strong> <div class="csl-block">Rose Oughtred, Jennifer Rust, Christie Chang, Bobby‐Joe Breitkreutz, Chris Stark, Andrew Willems, Lorrie Boucher, Genie Leung, Nadine Kolas, Frederick Zhang, … Mike Tyers</div> <em>Protein Science</em> (2020-11-23) <a href="https://doi.org/gk7skv">https://doi.org/gk7skv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1002/pro.3978">10.1002/pro.3978</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33070389">33070389</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7737760">PMC7737760</a></div></div>
</div>
<div id="ref-1663S4mH8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">48. </div><div class="csl-right-inline"><strong>The STRING database in 2021: customizable protein–protein networks, and functional characterization of user-uploaded gene/measurement sets</strong> <div class="csl-block">Damian Szklarczyk, Annika L Gable, Katerina C Nastou, David Lyon, Rebecca Kirsch, Sampo Pyysalo, Nadezhda T Doncheva, Marc Legeay, Tao Fang, Peer Bork, … Christian von Mering</div> <em>Nucleic Acids Research</em> (2020-11-25) <a href="https://doi.org/gh77tp">https://doi.org/gh77tp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkaa1074">10.1093/nar/gkaa1074</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33237311">33237311</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7779004">PMC7779004</a></div></div>
</div>
<div id="ref-eyEtjj2p" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">49. </div><div class="csl-right-inline"><strong>Cancer molecular markers: A guide to cancer detection and management</strong> <div class="csl-block">Meera Nair, Sardul Singh Sandhu, Anil Kumar Sharma</div> <em>Seminars in Cancer Biology</em> (2018-10) <a href="https://doi.org/gfdkxn">https://doi.org/gfdkxn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.semcancer.2018.02.002">10.1016/j.semcancer.2018.02.002</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29428478">29428478</a></div></div>
</div>
<div id="ref-1Hew9xuhZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">50. </div><div class="csl-right-inline"><strong>Support systems to guide clinical decision-making in precision oncology: The Cancer Core Europe Molecular Tumor Board Portal</strong> <div class="csl-block">David Tamborero, Rodrigo Dienstmann, Maan Haj Rachid, Jorrit Boekel, Richard Baird, Irene Braña, Luigi De Petris, Jeffrey Yachnin, Christophe Massard, Frans L Opdam, … </div> <em>Nature Medicine</em> (2020-07) <a href="https://doi.org/gh9htr">https://doi.org/gh9htr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41591-020-0969-2">10.1038/s41591-020-0969-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32632195">32632195</a></div></div>
</div>
<div id="ref-Prcc1PJY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">51. </div><div class="csl-right-inline"><strong>Cancer Genome Interpreter annotates the biological and clinical relevance of tumor alterations</strong> <div class="csl-block">David Tamborero, Carlota Rubio-Perez, Jordi Deu-Pons, Michael P Schroeder, Ana Vivancos, Ana Rovira, Ignasi Tusquets, Joan Albanell, Jordi Rodon, Josep Tabernero, … Nuria Lopez-Bigas</div> <em>Genome Medicine</em> (2018-03-28) <a href="https://doi.org/gmhdr5">https://doi.org/gmhdr5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s13073-018-0531-8">10.1186/s13073-018-0531-8</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29592813">29592813</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5875005">PMC5875005</a></div></div>
</div>
<div id="ref-xiGrNkR1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">52. </div><div class="csl-right-inline"><strong>The Sequence Ontology: a tool for the unification of genome annotations</strong> <div class="csl-block">Karen Eilbeck, Suzanna E Lewis, Christopher J Mungall, Mark Yandell, Lincoln Stein, Richard Durbin, Michael Ashburner</div> <em>Genome Biology</em> (2005-04-29) <a href="https://doi.org/frzhf4">https://doi.org/frzhf4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/gb-2005-6-5-r44">10.1186/gb-2005-6-5-r44</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/15892872">15892872</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1175956">PMC1175956</a></div></div>
</div>
<div id="ref-ez55Zlmu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">53. </div><div class="csl-right-inline"><strong>The Human Disease Ontology 2022 update</strong> <div class="csl-block">Lynn M Schriml, James B Munro, Mike Schor, Dustin Olley, Carrie McCracken, Victor Felix, J Allen Baron, Rebecca Jackson, Susan M Bello, Cynthia Bearer, … Carol Greene</div> <em>Nucleic Acids Research</em> (2021-11-10) <a href="https://doi.org/gsb57x">https://doi.org/gsb57x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab1063">10.1093/nar/gkab1063</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34755882">34755882</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728220">PMC8728220</a></div></div>
</div>
<div id="ref-nDiR3rcY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">54. </div><div class="csl-right-inline"><strong>OncoTree: A Cancer Classification System for Precision Oncology</strong> <div class="csl-block">Ritika Kundra, Hongxin Zhang, Robert Sheridan, Sahussapont Joseph Sirintrapun, Avery Wang, Angelica Ochoa, Manda Wilson, Benjamin Gross, Yichao Sun, Ramyasree Madupuri, … Nikolaus Schultz</div> <em>JCO Clinical Cancer Informatics</em> (2021-12) <a href="https://doi.org/gmgf8m">https://doi.org/gmgf8m</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1200/cci.20.00108">10.1200/cci.20.00108</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33625877">33625877</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8240791">PMC8240791</a></div></div>
</div>
<div id="ref-11HFBpGej" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">55. </div><div class="csl-right-inline"><strong>Open Targets: a platform for therapeutic target identification and validation</strong> <div class="csl-block">Gautier Koscielny, Peter An, Denise Carvalho-Silva, Jennifer A Cham, Luca Fumis, Rippa Gasparyan, Samiul Hasan, Nikiforos Karamanis, Michael Maguire, Eliseo Papa, … Ian Dunham</div> <em>Nucleic Acids Research</em> (2016-11-29) <a href="https://doi.org/f9v9hf">https://doi.org/f9v9hf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkw1055">10.1093/nar/gkw1055</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27899665">27899665</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210543">PMC5210543</a></div></div>
</div>
<div id="ref-9fcv9d5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">56. </div><div class="csl-right-inline"><strong>Network expansion of genetic associations defines a pleiotropy map of human cell biology</strong> <div class="csl-block">Inigo Barrio-Hernandez, Jeremy Schwartzentruber, Anjali Shrivastava, Noemi del-Toro, Qian Zhang, Glyn Bradley, Henning Hermjakob, Sandra Orchard, Ian Dunham, Carl A Anderson, … Pedro Beltrao</div> <em>Cold Spring Harbor Laboratory</em> (2021-07-19) <a href="https://doi.org/gs52mr">https://doi.org/gs52mr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2021.07.19.452924">10.1101/2021.07.19.452924</a></div></div>
</div>
<div id="ref-iDcPtWv1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">57. </div><div class="csl-right-inline"><strong>GRAPE for Fast and Scalable Graph Processing and random walk-based Embedding</strong> <div class="csl-block">Luca Cappelletti, Tommaso Fontana, Elena Casiraghi, Vida Ravanmehr, Tiffany J Callahan, Carlos Cano, Marcin P Joachimiak, Christopher J Mungall, Peter N Robinson, Justin Reese, Giorgio Valentini</div> <em>arXiv</em> (2021) <a href="https://doi.org/gs52m8">https://doi.org/gs52m8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2110.06196">10.48550/arxiv.2110.06196</a></div></div>
</div>
<div id="ref-Tm14Ibs7" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">58. </div><div class="csl-right-inline"><strong>GEARS: Predicting transcriptional outcomes of novel multi-gene perturbations</strong> <div class="csl-block">Yusuf Roohani, Kexin Huang, Jure Leskovec</div> <em>Cold Spring Harbor Laboratory</em> (2022-07-14) <a href="https://doi.org/gqhrzs">https://doi.org/gqhrzs</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.07.12.499735">10.1101/2022.07.12.499735</a></div></div>
</div>
<div id="ref-11A8WJC5q" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">59. </div><div class="csl-right-inline"><strong>On the Privacy of Federated Pipelines</strong> <div class="csl-block">Reza Nasirigerdeh, Reihaneh Torkzadehmahani, Jan Baumbach, David B Blumenthal</div> <em>Proceedings of the 44th International ACM SIGIR Conference on Research and Development in Information Retrieval</em> (2021-07-11) <a href="https://doi.org/gs52mx">https://doi.org/gs52mx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/3404835.3462996">10.1145/3404835.3462996</a></div></div>
</div>
<div id="ref-aDwLmWW9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">60. </div><div class="csl-right-inline"><strong>The FeatureCloud AI Store for Federated Learning in Biomedicine and Beyond</strong> <div class="csl-block">Julian Matschinske, Julian Späth, Reza Nasirigerdeh, Reihaneh Torkzadehmahani, Anne Hartebrodt, Balázs Orbán, Sándor Fejér, Olga Zolotareva, Mohammad Bakhtiari, Béla Bihari, … Jan Baumbach</div> <em>arXiv</em> (2021) <a href="https://doi.org/gs52m6">https://doi.org/gs52m6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2105.05734">10.48550/arxiv.2105.05734</a></div></div>
</div>
<div id="ref-19oj9JJD5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">61. </div><div class="csl-right-inline"><strong>[Orphanet: a European database for rare diseases].</strong> <div class="csl-block">SS Weinreich, R Mangon, JJ Sikkens, ME en Teeuw, MC Cornel</div> <em>Nederlands tijdschrift voor geneeskunde</em> (2008-03-01) <a href="https://www.ncbi.nlm.nih.gov/pubmed/18389888">https://www.ncbi.nlm.nih.gov/pubmed/18389888</a> <div class="csl-block">PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/18389888">18389888</a></div></div>
</div>
<div id="ref-nELyJbhp" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">62. </div><div class="csl-right-inline"><strong>German approach of coding rare diseases with ICD-10-GM and Orpha numbers in routine settings</strong> <div class="csl-block">Stefanie Weber, Magdalena Dávila</div> <em>Orphanet Journal of Rare Diseases</em> (2014) <a href="https://doi.org/gs52mz">https://doi.org/gs52mz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/1750-1172-9-s1-o10">10.1186/1750-1172-9-s1-o10</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4249588">PMC4249588</a></div></div>
</div>
<div id="ref-o7nDQFO1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">63. </div><div class="csl-right-inline"><strong>The Human Phenotype Ontology</strong> <div class="csl-block">PN Robinson, S Mundlos</div> <em>Clinical Genetics</em> (2010-05-09) <a href="https://doi.org/cj87nn">https://doi.org/cj87nn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1399-0004.2010.01436.x">10.1111/j.1399-0004.2010.01436.x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20412080">20412080</a></div></div>
</div>
<div id="ref-6uGIWm7L" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">64. </div><div class="csl-right-inline"><strong>SNOMED-CT: The advanced terminology and coding system for eHealth.</strong> <div class="csl-block">Kevin Donnelly</div> <em>Studies in health technology and informatics</em> (2006) <a href="https://www.ncbi.nlm.nih.gov/pubmed/17095826">https://www.ncbi.nlm.nih.gov/pubmed/17095826</a> <div class="csl-block">PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17095826">17095826</a></div></div>
</div>
<div id="ref-yKZ8alU5" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">65. </div><div class="csl-right-inline"><strong>MIRACUM: Medical Informatics in Research and Care in University Medicine</strong> <div class="csl-block">Hans-Ulrich Prokosch, Till Acker, Johannes Bernarding, Harald Binder, Martin Boeker, Melanie Boerries, Philipp Daumke, Thomas Ganslandt, Jürgen Hesser, Gunther Höning, … Holger Storf</div> <em>Methods of Information in Medicine</em> (2018-07) <a href="https://doi.org/gdzpqv">https://doi.org/gdzpqv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3414/me17-02-0025">10.3414/me17-02-0025</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30016814">30016814</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6178200">PMC6178200</a></div></div>
</div>
<div id="ref-eFQyj5Wh" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">66. </div><div class="csl-right-inline"><a href="https://www.medizininformatik-initiative.de/sites/default/files/2018-07/2018-03_mdi_Der%20Kerndatensatz%20der%20Medizininformatik-Initiative%20Ein%20Schritt%20zur%20Sekund%C3%A4rnutzung%20von%20Versorgungsdaten%20auf%20nationaler%20Ebene.pdf">https://www.medizininformatik-initiative.de/sites/default/files/2018-07/2018-03_mdi_Der%20Kerndatensatz%20der%20Medizininformatik-Initiative%20Ein%20Schritt%20zur%20Sekund%C3%A4rnutzung%20von%20Versorgungsdaten%20auf%20nationaler%20Ebene.pdf</a></div>
</div>
<div id="ref-lrjMS8FK" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">67. </div><div class="csl-right-inline"><strong>Study of Health in Pomerania (SHIP): A health examination survey in an east German region: Objectives and design</strong> <div class="csl-block">Ulrich John, Elke Hensel, Jan L�demann, Marion Piek, Sybille Sauer, Christiane Adam, Gabriele Born, Dietrich Alte, Eberhard Greiser, Ursula Haertel, … Christof Kessler</div> <em>Sozial- und Pr�ventivmedizin SPM</em> (2001-05) <a href="https://doi.org/c4gt9k">https://doi.org/c4gt9k</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/bf01324255">10.1007/bf01324255</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/11565448">11565448</a></div></div>
</div>
<div id="ref-u1j8EALG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">68. </div><div class="csl-right-inline"><strong>Cohort profile: Greifswald approach to individualized medicine (GANI_MED)</strong> <div class="csl-block">Hans J Grabe, Heinrich Assel, Thomas Bahls, Marcus Dörr, Karlhans Endlich, Nicole Endlich, Pia Erdmann, Ralf Ewert, Stephan B Felix, Beate Fiene, … Heyo K Kroemer</div> <em>Journal of Translational Medicine</em> (2014) <a href="https://doi.org/gpf2mh">https://doi.org/gpf2mh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/1479-5876-12-144">10.1186/1479-5876-12-144</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24886498">24886498</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4040487">PMC4040487</a></div></div>
</div>
<div id="ref-18dC6FMrj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">69. </div><div class="csl-right-inline"><strong>BioPortal: ontologies and integrated data resources at the click of a mouse</strong> <div class="csl-block">NF Noy, NH Shah, PL Whetzel, B Dai, M Dorf, N Griffith, C Jonquet, DL Rubin, M-A Storey, CG Chute, MA Musen</div> <em>Nucleic Acids Research</em> (2009-05-29) <a href="https://doi.org/dm869h">https://doi.org/dm869h</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkp440">10.1093/nar/gkp440</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19483092">19483092</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2703982">PMC2703982</a></div></div>
</div>
<div id="ref-nC3lHM8C" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">70. </div><div class="csl-right-inline"><strong>The German Corona Consensus Dataset (GECCO): a standardized dataset for COVID-19 research in university medicine and beyond</strong> <div class="csl-block">Julian Sass, Alexander Bartschke, Moritz Lehne, Andrea Essenwanger, Eugenia Rinaldi, Stefanie Rudolph, Kai U Heitmann, Jörg J Vehreschild, Christof von Kalle, Sylvia Thun</div> <em>BMC Medical Informatics and Decision Making</em> (2020-12) <a href="https://doi.org/gs52m2">https://doi.org/gs52m2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s12911-020-01374-w">10.1186/s12911-020-01374-w</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33349259">33349259</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7751265">PMC7751265</a></div></div>
</div>
<div id="ref-QrLcJJPA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">71. </div><div class="csl-right-inline"><strong>CovidGraph: a graph to fight COVID-19</strong> <div class="csl-block">Lea Gütebier, Tim Bleimehl, Ron Henkel, Jamie Munro, Sebastian Müller, Axel Morgner, Jakob Laenge, Anke Pachauer, Alexander Erdl, Jens Weimar, … Alexander Jarasch</div> <em>Bioinformatics</em> (2022-08-30) <a href="https://doi.org/gs52mn">https://doi.org/gs52mn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/bioinformatics/btac592">10.1093/bioinformatics/btac592</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36040169">36040169</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9563682">PMC9563682</a></div></div>
</div>
<div id="ref-16z4LZSdJ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">72. </div><div class="csl-right-inline"><strong>Sherlock: an open-source data platform to store, analyze and integrate Big Data for computational biologists</strong> <div class="csl-block">Balazs Bohar, David Fazekas, Matthew Madgwick, Luca Csabai, Marton Olbei, Tamás Korcsmáros, Mate Szalay-Beko</div> <em>F1000Research</em> (2023-01-12) <a href="https://doi.org/gs52m4">https://doi.org/gs52m4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.12688/f1000research.52791.3">10.12688/f1000research.52791.3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36533093">36533093</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9731172">PMC9731172</a></div></div>
</div>
<div id="ref-HcpxNvpk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">73. </div><div class="csl-right-inline"><strong>Deciphering cell–cell interactions and communication from gene expression</strong> <div class="csl-block">Erick Armingol, Adam Officer, Olivier Harismendy, Nathan E Lewis</div> <em>Nature Reviews Genetics</em> (2020-11-09) <a href="https://doi.org/ghjj3h">https://doi.org/ghjj3h</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41576-020-00292-x">10.1038/s41576-020-00292-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33168968">33168968</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7649713">PMC7649713</a></div></div>
</div>
<div id="ref-1A6n4edrt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">74. </div><div class="csl-right-inline"><strong>The landscape of cell–cell communication through single-cell transcriptomics</strong> <div class="csl-block">Axel A Almet, Zixuan Cang, Suoqin Jin, Qing Nie</div> <em>Current Opinion in Systems Biology</em> (2021-06) <a href="https://doi.org/gs52mh">https://doi.org/gs52mh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.coisb.2021.03.007">10.1016/j.coisb.2021.03.007</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33969247">33969247</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8104132">PMC8104132</a></div></div>
</div>
<div id="ref-1EXXNpykM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">75. </div><div class="csl-right-inline"><strong>MEBOCOST: Metabolite-mediated Cell Communication Modeling by Single Cell Transcriptome</strong> <div class="csl-block">Rongbin Zheng, Yang Zhang, Tadataka Tsuji, Xinlei Gao, Allon Wagner, Nir Yosef, Hong Chen, Lili Zhang, Yu-Hua Tseng, Kaifu Chen</div> <em>Cold Spring Harbor Laboratory</em> (2022-05-31) <a href="https://doi.org/gsw74q">https://doi.org/gsw74q</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1101/2022.05.30.494067">10.1101/2022.05.30.494067</a></div></div>
</div>
<div id="ref-PSwd2RWP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">76. </div><div class="csl-right-inline"><strong>Inferring neuron-neuron communications from single-cell transcriptomics through NeuronChat</strong> <div class="csl-block">Wei Zhao, Kevin G Johnston, Honglei Ren, Xiangmin Xu, Qing Nie</div> <em>Nature Communications</em> (2023-02-28) <a href="https://doi.org/gsx7kz">https://doi.org/gsx7kz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-023-36800-w">10.1038/s41467-023-36800-w</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36854676">36854676</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9974942">PMC9974942</a></div></div>
</div>
<div id="ref-1FhKLN5iq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">77. </div><div class="csl-right-inline"><strong>Single-cell roadmap of human gonadal development</strong> <div class="csl-block">Luz Garcia-Alonso, Valentina Lorenzi, Cecilia Icoresi Mazzeo, João Pedro Alves-Lopes, Kenny Roberts, Carmen Sancho-Serra, Justin Engelbert, Magda Marečková, Wolfram H Gruhn, Rachel A Botting, … Roser Vento-Tormo</div> <em>Nature</em> (2022-07-06) <a href="https://doi.org/gqgf9r">https://doi.org/gqgf9r</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41586-022-04918-4">10.1038/s41586-022-04918-4</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/35794482">35794482</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9300467">PMC9300467</a></div></div>
</div>
<div id="ref-aNFFB1Q" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">78. </div><div class="csl-right-inline"><strong>Comparison of methods and resources for cell-cell communication inference from single-cell RNA-Seq data</strong> <div class="csl-block">Daniel Dimitrov, Dénes Türei, Martin Garrido-Rodriguez, Paul L Burmedi, James S Nagai, Charlotte Boys, Ricardo O Ramirez Flores, Hyojin Kim, Bence Szalai, Ivan G Costa, … Julio Saez-Rodriguez</div> <em>Nature Communications</em> (2022-06-09) <a href="https://doi.org/grkzjh">https://doi.org/grkzjh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-022-30755-0">10.1038/s41467-022-30755-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/35680885">35680885</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9184522">PMC9184522</a></div></div>
</div>
<div id="ref-VwS8W9dk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">79. </div><div class="csl-right-inline"><strong>KEGG: new perspectives on genomes, pathways, diseases and drugs</strong> <div class="csl-block">Minoru Kanehisa, Miho Furumichi, Mao Tanabe, Yoko Sato, Kanae Morishima</div> <em>Nucleic Acids Research</em> (2016-11-28) <a href="https://doi.org/f9v6kv">https://doi.org/f9v6kv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkw1092">10.1093/nar/gkw1092</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27899662">27899662</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5210567">PMC5210567</a></div></div>
</div>
<div id="ref-Gout7oWo" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">80. </div><div class="csl-right-inline"><strong>The reactome pathway knowledgebase 2022</strong> <div class="csl-block">Marc Gillespie, Bijay Jassal, Ralf Stephan, Marija Milacic, Karen Rothfels, Andrea Senff-Ribeiro, Johannes Griss, Cristoffer Sevilla, Lisa Matthews, Chuqiao Gong, … Peter D’Eustachio</div> <em>Nucleic Acids Research</em> (2021-11-12) <a href="https://doi.org/gpm2r5">https://doi.org/gpm2r5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab1028">10.1093/nar/gkab1028</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34788843">34788843</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8689983">PMC8689983</a></div></div>
</div>
<div id="ref-v9JUscZ1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">81. </div><div class="csl-right-inline"><strong>Rhea, the reaction knowledgebase in 2022</strong> <div class="csl-block">Parit Bansal, Anne Morgat, Kristian B Axelsen, Venkatesh Muthukrishnan, Elisabeth Coudert, Lucila Aimo, Nevila Hyka-Nouspikel, Elisabeth Gasteiger, Arnaud Kerhornou, Teresa Batista Neto, … Alan Bridge</div> <em>Nucleic Acids Research</em> (2021-11-10) <a href="https://doi.org/gqk8b7">https://doi.org/gqk8b7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab1016">10.1093/nar/gkab1016</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34755880">34755880</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728268">PMC8728268</a></div></div>
</div>
<div id="ref-Rygf3Cgf" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">82. </div><div class="csl-right-inline"><strong>HMDB 5.0: the Human Metabolome Database for 2022</strong> <div class="csl-block">David S Wishart, AnChi Guo, Eponine Oler, Fei Wang, Afia Anjum, Harrison Peters, Raynard Dizon, Zinat Sayeeda, Siyang Tian, Brian L Lee, … Vasuk Gautam</div> <em>Nucleic Acids Research</em> (2021-11-19) <a href="https://doi.org/gpfs92">https://doi.org/gpfs92</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab1062">10.1093/nar/gkab1062</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34986597">34986597</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8728138">PMC8728138</a></div></div>
</div>
<div id="ref-TYquyrUR" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">83. </div><div class="csl-right-inline"><strong>Recon3D enables a three-dimensional view of gene variation in human metabolism</strong> <div class="csl-block">Elizabeth Brunk, Swagatika Sahoo, Daniel C Zielinski, Ali Altunkaya, Andreas Dräger, Nathan Mih, Francesco Gatto, Avlant Nilsson, German Andres Preciat Gonzalez, Maike Kathrin Aurich, … Bernhard O Palsson</div> <em>Nature Biotechnology</em> (2018-02-19) <a href="https://doi.org/gg7wf5">https://doi.org/gg7wf5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nbt.4072">10.1038/nbt.4072</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29457794">29457794</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5840010">PMC5840010</a></div></div>
</div>
<div id="ref-QANpB4kr" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">84. </div><div class="csl-right-inline"><strong>An atlas of human metabolism</strong> <div class="csl-block">Jonathan L Robinson, Pınar Kocabaş, Hao Wang, Pierre-Etienne Cholley, Daniel Cook, Avlant Nilsson, Mihail Anton, Raphael Ferreira, Iván Domenzain, Virinchi Billa, … Jens Nielsen</div> <em>Science Signaling</em> (2020-03-24) <a href="https://doi.org/ggqgs6">https://doi.org/ggqgs6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1126/scisignal.aaz1482">10.1126/scisignal.aaz1482</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32209698">32209698</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7331181">PMC7331181</a></div></div>
</div>
<div id="ref-bc1DWCco" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">85. </div><div class="csl-right-inline"><strong>STITCH 5: augmenting protein–chemical interaction networks with tissue and affinity data</strong> <div class="csl-block">Damian Szklarczyk, Alberto Santos, Christian von Mering, Lars Juhl Jensen, Peer Bork, Michael Kuhn</div> <em>Nucleic Acids Research</em> (2015-11-20) <a href="https://doi.org/f8cprg">https://doi.org/f8cprg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkv1277">10.1093/nar/gkv1277</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26590256">26590256</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4702904">PMC4702904</a></div></div>
</div>
<div id="ref-1GjHUK5qj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">86. </div><div class="csl-right-inline"><strong>The IUPHAR/BPS guide to PHARMACOLOGY in 2022: curating pharmacology for COVID-19, malaria and antibacterials</strong> <div class="csl-block">Simon D Harding, Jane F Armstrong, Elena Faccenda, Christopher Southan, Stephen PH Alexander, Anthony P Davenport, Adam J Pawson, Michael Spedding, Jamie A Davies</div> <em>Nucleic Acids Research</em> (2021-10-30) <a href="https://doi.org/gs52mp">https://doi.org/gs52mp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkab1010">10.1093/nar/gkab1010</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34718737">34718737</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8689838">PMC8689838</a></div></div>
</div>
<div id="ref-GBpX8fvk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">87. </div><div class="csl-right-inline"><strong>Protein-metabolite interactomics of carbohydrate metabolism reveal regulation of lactate dehydrogenase</strong> <div class="csl-block">Kevin G Hicks, Ahmad A Cluntun, Heidi L Schubert, Sean R Hackett, Jordan A Berg, Paul G Leonard, Mariana A Ajalla Aleixo, Youjia Zhou, Alex J Bott, Sonia R Salvatore, … Jared Rutter</div> <em>Science</em> (2023-03-10) <a href="https://doi.org/gsb7t6">https://doi.org/gsb7t6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1126/science.abm3452">10.1126/science.abm3452</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36893255">36893255</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10262665">PMC10262665</a></div></div>
</div>
<div id="ref-171uilzJ7" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">88. </div><div class="csl-right-inline"><strong>Towards a Rosetta stone for metabolomics: recommendations to overcome inconsistent metabolite nomenclature</strong> <div class="csl-block">Ville Koistinen, Olli Kärkkäinen, Pekka Keski-Rahkonen, Hiroshi Tsugawa, Augustin Scalbert, Masanori Arita, David Wishart, Kati Hanhineva</div> <em>Nature Metabolism</em> (2023-03-08) <a href="https://doi.org/gs52mm">https://doi.org/gs52mm</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s42255-023-00757-3">10.1038/s42255-023-00757-3</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36890347">36890347</a></div></div>
</div>
<div id="ref-117E2kGkQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">89. </div><div class="csl-right-inline"><strong>Consistency, Inconsistency, and Ambiguity of Metabolite Names in Biochemical Databases Used for Genome-Scale Metabolic Modelling</strong> <div class="csl-block">Nhung Pham, Ruben van Heck, Jesse van Dam, Peter Schaap, Edoardo Saccenti, Maria Suarez-Diez</div> <em>Metabolites</em> (2019-02-06) <a href="https://doi.org/gs52m5">https://doi.org/gs52m5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3390/metabo9020028">10.3390/metabo9020028</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30736318">30736318</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6409771">PMC6409771</a></div></div>
</div>
<div id="ref-RnnKuIPr" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">90. </div><div class="csl-right-inline"><strong>Integrated cross-study datasets of genetic dependencies in cancer</strong> <div class="csl-block">Clare Pacini, Joshua M Dempster, Isabella Boyle, Emanuel Gonçalves, Hanna Najgebauer, Emre Karakoc, Dieudonne van der Meer, Andrew Barthorpe, Howard Lightfoot, Patricia Jaaks, … Francesco Iorio</div> <em>Nature Communications</em> (2021-03-12) <a href="https://doi.org/gnpt56">https://doi.org/gnpt56</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41467-021-21898-7">10.1038/s41467-021-21898-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33712601">33712601</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7955067">PMC7955067</a></div></div>
</div>
<div id="ref-1GDvfevZN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">91. </div><div class="csl-right-inline"><strong>GenomicKB: a knowledge graph for the human genome</strong> <div class="csl-block">Fan Feng, Feitong Tang, Yijia Gao, Dongyu Zhu, Tianjun Li, Shuyuan Yang, Yuan Yao, Yuanhao Huang, Jie Liu</div> <em>Nucleic Acids Research</em> (2022-11-01) <a href="https://doi.org/gs52mq">https://doi.org/gs52mq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/nar/gkac957">10.1093/nar/gkac957</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/36318240">36318240</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9825430">PMC9825430</a></div></div>
</div>
<div id="ref-BvOaZVwT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">92. </div><div class="csl-right-inline"><strong>Scientific evidence based rare disease research discovery with research funding data in knowledge graph</strong> <div class="csl-block">Qian Zhu, Ðắc-Trung Nguyễn, Timothy Sheils, Gioconda Alyea, Eric Sid, Yanji Xu, James Dickens, Ewy A Mathé, Anne Pariser</div> <em>Orphanet Journal of Rare Diseases</em> (2021-11-18) <a href="https://doi.org/gs52m3">https://doi.org/gs52m3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/s13023-021-02120-9">10.1186/s13023-021-02120-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/34794473">34794473</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8600882">PMC8600882</a></div></div>
</div>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
